## Project Description

In a PUBG game, up to 100 players start in each match (matchId). Players can be on teams (groupId) which get ranked at the end of the game (winPlacePerc) based on how many other teams are still alive when they are eliminated. In game, players can pick up different munitions, revive downed-but-not-out (knocked) teammates, drive vehicles, swim, run, shoot, and experience all of the consequences -- such as falling too far or running themselves over and eliminating themselves.

This Project aims to predict players' finishing placement based on their final stats in a PUBG game. The data comes from matches of all types: solos, duos, squads, and custom; there is no guarantee of there being 100 players per match, nor at most 4 player per group.

## Part 1: Explore Data
From the .csv file import dataframe. The delimiter is comma and the first row is the header.

In [0]:
from pyspark.sql.types import *
# File location and type
file_location = "/FileStore/tables/train_V2-1.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# To increase the accuracy, we set all numerical values as doubleType
customSchema = StructType([ \
    StructField("Id", StringType(), True),
    StructField("groupId", StringType(), True),
    StructField("matchId", StringType(), True),
    StructField("assists", DoubleType(), True), \
    StructField("boosts", DoubleType(), True), \
    StructField("damageDealt", DoubleType(), True), \
    StructField("DBNOs", DoubleType(), True), \
    StructField("headshotKills", DoubleType(), True), \
    StructField("heals", DoubleType(), True), \
    StructField("killPlace", DoubleType(), True), \
    StructField("killPoints", StringType(), True),
    StructField("kills", DoubleType(), True), \
    StructField("killStreaks", DoubleType(), True), \
    StructField("longestKill", DoubleType(), True),
    StructField("matchDuration", DoubleType(), True),
    StructField("matchType", StringType(), True),
    StructField("maxPlace", DoubleType(), True),
    StructField("numGroups", DoubleType(), True),
    StructField("rankPoints", DoubleType(), True),
    StructField("revives", DoubleType(), True), \
    StructField("rideDistance", DoubleType(), True), \
    StructField("roadKills", DoubleType(), True), \
    StructField("swimDistance", DoubleType(), True), \
    StructField("teamKills", DoubleType(), True),
    StructField("vehicleDestroys", DoubleType(), True), \
    StructField("walkDistance", DoubleType(), True), \
    StructField("weaponsAcquired", DoubleType(), True),
    StructField("winPoints", DoubleType(), True),
    StructField("winPlacePerc", DoubleType(), True)])

# The applied options are for CSV files. For other file types, these will be ignored.
originDF = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location, schema = customSchema)

display(originDF)

Id groupId matchId assists boosts damageDealt DBNOs headshotKills heals killPlace killPoints kills killStreaks longestKill matchDuration matchType maxPlace numGroups rankPoints revives rideDistance roadKills swimDistance teamKills vehicleDestroys walkDistance weaponsAcquired winPoints winPlacePerc 7f96b2f878858a 4d4b580de459be a10357fd1a4a91 0.0 0.0 0.0 0.0 0.0 0.0 60.0 1241 0.0 0.0 0.0 1306.0 squad-fpp 28.0 26.0 -1.0 0.0 0.0 0.0 0.0 0.0 0.0 244.8 1.0 1466.0 0.4444 eef90569b9d03c 684d5656442f9e aeb375fc57110c 0.0 0.0 91.47 0.0 0.0 0.0 57.0 0 0.0 0.0 0.0 1777.0 squad-fpp 26.0 25.0 1484.0 0.0 0.0045 0.0 11.04 0.0 0.0 1434.0 5.0 0.0 0.64 1eaf90ac73de72 6a4a42c3245a74 110163d8bb94ae 1.0 0.0 68.0 0.0 0.0 0.0 47.0 0 0.0 0.0 0.0 1318.0 duo 50.0 47.0 1491.0 0.0 0.0 0.0 0.0 0.0 0.0 161.8 2.0 0.0 0.7755 4616d365dd2853 a930a9c79cd721 f1f1f4ef412d7e 0.0 0.0 32.9 0.0 0.0 0.0 75.0 0 0.0 0.0 0.0 1436.0 squad-fpp 31.0 30.0 1408.0 0.0 0.0 0.0 0.0 0.0 0.0 202.7 3.0 0.0 0.1667 315c96c26c9aac de04010b3458dd 6dc8ff871e21e6 0.0 0.0 100.0 0.0 0.0 0.0 45.0 0 1.0 1.0 58.53 1424.0 solo-fpp 97.0 95.0 1560.0 0.0 0.0 0.0 0.0 0.0 0.0 49.75 2.0 0.0 0.1875 ff79c12f326506 289a6836a88d27 bac52627a12114 0.0 0.0 100.0 1.0 1.0 0.0 44.0 0 1.0 1.0 18.44 1395.0 squad-fpp 28.0 28.0 1418.0 0.0 0.0 0.0 0.0 0.0 0.0 34.7 1.0 0.0 0.037 95959be0e21ca3 2c485a1ad3d0f1 a8274e903927a2 0.0 0.0 0.0 0.0 0.0 0.0 96.0 1262 0.0 0.0 0.0 1316.0 squad-fpp 28.0 28.0 -1.0 0.0 0.0 0.0 0.0 0.0 0.0 13.5 1.0 1497.0 0.0 311b84c6ff4390 eaba5fcb7fc1ae 292611730ca862 0.0 0.0 8.538 0.0 0.0 0.0 48.0 1000 0.0 0.0 0.0 1967.0 solo-fpp 96.0 92.0 -1.0 0.0 2004.0 0.0 0.0 0.0 0.0 1089.0 6.0 1500.0 0.7368 1a68204ccf9891 47cfbb04e1b1a2 df014fbee741c6 0.0 0.0 51.6 0.0 0.0 0.0 64.0 0 0.0 0.0 0.0 1375.0 squad 28.0 27.0 1493.0 0.0 0.0 0.0 0.0 0.0 0.0 799.9 4.0 0.0 0.3704 e5bb5a43587253 759bb6f7514fd2 3d3031c795305b 0.0 0.0 37.27 0.0 0.0 0.0 74.0 0 0.0 0.0 0.0 1930.0 squad 29.0 27.0 1349.0 0.0 0.0 0.0 0.0 0.0 0.0 65.67 1.0 0.0 0.2143 2b574d43972813 c549efede67ad3 2dd6ddb8320fc1 0.0 0.0 28.38 0.0 0.0 0.0 75.0 0 0.0 0.0 0.0 1811.0 squad-fpp 29.0 29.0 1475.0 0.0 0.0 0.0 0.0 0.0 0.0 868.3 9.0 0.0 0.3929 8de328a74658a9 f643df9df3877c 80170383d90003 0.0 0.0 137.9 1.0 0.0 0.0 64.0 0 0.0 0.0 0.0 1384.0 duo-fpp 48.0 46.0 1488.0 0.0 0.0 0.0 0.0 0.0 0.0 451.7 1.0 0.0 0.4043 ce4f6ac165705e da24cdb91969cc 535b5dbd965a94 0.0 0.0 0.0 0.0 0.0 0.0 37.0 0 0.0 0.0 0.0 1774.0 squad-fpp 29.0 28.0 1766.0 0.0 6639.0 0.0 0.0 0.0 0.0 2784.0 6.0 0.0 0.9286 b7807186e3f679 3c08e461874749 2c30ddf481c52d 0.0 1.0 324.2 0.0 1.0 5.0 5.0 986 4.0 1.0 49.83 1886.0 solo-fpp 97.0 94.0 -1.0 0.0 1228.0 0.0 76.84 0.0 0.0 2050.0 6.0 1462.0 0.875 8e244ac61b6aab d40d0c7d3573a1 94e1c1cc443c65 0.0 1.0 122.8 1.0 0.0 2.0 25.0 1411 1.0 1.0 37.91 1458.0 squad-fpp 31.0 30.0 -1.0 1.0 1237.0 0.0 60.29 0.0 0.0 1666.0 5.0 1531.0 0.9 12d8d4bd94312c fe52d481bae68b 6fd9e765ddd0c5 0.0 0.0 80.71 1.0 0.0 0.0 72.0 0 0.0 0.0 0.0 1313.0 duo-fpp 48.0 48.0 1495.0 1.0 0.0 0.0 0.0 0.0 0.0 105.1 5.0 0.0 0.2766 62f2f0917d84b2 f61b698274d9f5 1d6cfe0f6f23b0 0.0 2.0 81.71 1.0 0.0 14.0 25.0 0 1.0 1.0 9.158 1882.0 squad-fpp 27.0 26.0 1436.0 0.0 519.9 0.0 0.0 0.0 0.0 3674.0 7.0 0.0 0.7308 92022479b92ce7 2f2c33f548c4b9 07948d723b9c0f 0.0 3.0 254.3 0.0 0.0 12.0 13.0 0 2.0 1.0 36.0 1371.0 solo-fpp 96.0 95.0 1536.0 0.0 2367.0 0.0 15.29 0.0 0.0 1787.0 3.0 0.0 0.8211 7bd224781f064b 6dde607d151819 733af30cc00099 0.0 0.0 0.0 0.0 0.0 0.0 79.0 0 0.0 0.0 0.0 1301.0 squad-fpp 27.0 26.0 1355.0 0.0 0.0 0.0 0.0 0.0 0.0 137.4 2.0 0.0 0.1923 71cbdbc3b263e5 7b61f74b51906c a329ac99449ad7 0.0 1.0 65.28 0.0 0.0 1.0 48.0 1349 0.0 0.0 0.0 1322.0 squad-fpp 30.0 28.0 0.0 0.0 0.0 0.0 20.85 0.0 0.0 3310.0 3.0 1479.0 0.931 02ace8c6e58461 a4bc548028f800 80f2b8448e474b 0.0 4.0 269.1 0.0 1.0 8.0 18.0 0 2.0 1.0 7.438 1890.0 duo-fpp 48.0 48.0 1491.0 1.0 2734.0 0.0 0.0 0.0 0.0 1794.0 5.0 0.0 0.6383 00341b1caa5420 d661a2d19e7ae9 f3956286eb39a5 0.0 0.0 158.7 1.0 0.0 0.0 75.0 0 0.0 0.0 0.0 1853.0 squad-fpp 29.0 28.0 1522.

Check the type of all data. Notice that some of them are double and string instead of int.

In [0]:
print originDF.dtypes

[('Id', 'string'), ('groupId', 'string'), ('matchId', 'string'), ('assists', 'double'), ('boosts', 'double'), ('damageDealt', 'double'), ('DBNOs', 'double'), ('headshotKills', 'double'), ('heals', 'double'), ('killPlace', 'double'), ('killPoints', 'string'), ('kills', 'double'), ('killStreaks', 'double'), ('longestKill', 'double'), ('matchDuration', 'double'), ('matchType', 'string'), ('maxPlace', 'double'), ('numGroups', 'double'), ('rankPoints', 'double'), ('revives', 'double'), ('rideDistance', 'double'), ('roadKills', 'double'), ('swimDistance', 'double'), ('teamKills', 'double'), ('vehicleDestroys', 'double'), ('walkDistance', 'double'), ('weaponsAcquired', 'double'), ('winPoints', 'double'), ('winPlacePerc', 'double')]

## Part 2: Preliminary Analyses
Then we will do some original analyses of the dataset.

**Schema Definition**

Here's the schema definition:

 - DBNOs - Number of enemy players knocked.
 - assists - Number of enemy players this player damaged that were killed by teammates.
 - boosts - Number of boost items used.
 - damageDealt - Total damage dealt. Note: Self inflicted damage is subtracted.
 - headshotKills - Number of enemy players killed with headshots.
 - heals - Number of healing items used.
 - Id - Player’s Id
 - killPlace - Ranking in match of number of enemy players killed.
 - killPoints - Kills-based external ranking of player. (Think of this as an Elo ranking where only kills matter.) If there is a value other than -1 in rankPoints, then any 0 in killPoints should be treated as a “None”.
 - killStreaks - Max number of enemy players killed in a short amount of time.
 - kills - Number of enemy players killed.
 - longestKill - Longest distance between player and player killed at time of death. This may be misleading, as downing a player and driving away may lead to a large longestKill stat.
 - matchDuration - Duration of match in seconds.
 - matchId - ID to identify match. There are no matches that are in both the training and testing set.
 - matchType - String identifying the game mode that the data comes from. The standard modes are “solo”, “duo”, “squad”, “solo-fpp”, “duo-fpp”, and “squad-fpp”; other modes are from events or custom matches.
 - rankPoints - Elo-like ranking of player. This ranking is inconsistent and is being deprecated in the API’s next version, so use with caution. Value of -1 takes place of “None”.
 - revives - Number of times this player revived teammates.
 - rideDistance - Total distance traveled in vehicles measured in meters.
 - roadKills - Number of kills while in a vehicle.
 - swimDistance - Total distance traveled by swimming measured in meters.
 - teamKills - Number of times this player killed a teammate.
 - vehicleDestroys - Number of vehicles destroyed.
 - walkDistance - Total distance traveled on foot measured in meters.
 - weaponsAcquired - Number of weapons picked up.
 - winPoints - Win-based external ranking of player. (Think of this as an Elo ranking where only winning matters.) If there is a value other than -1 in rankPoints, then any 0 in winPoints should be treated as a “None”.
 - groupId - ID to identify a group within a match. If the same group of players plays in different matches, they will have a different groupId each time.
 - numGroups - Number of groups we have data for in the match.
 - maxPlace - Worst placement we have data for in the match. This may not match with numGroups, as sometimes the data skips over placements.
 - winPlacePerc - The target of prediction. This is a percentile winning placement, where 1 corresponds to 1st place, and 0 corresponds to last place in the match. It is calculated off of maxPlace, not numGroups, so it is possible to have missing chunks in a match

**Notice that some of the schema is not related to the winPlace, so we remove them right away.** 

 - Id - Player’s Id
 - groupId - ID to identify a group within a match. If the same group of players plays in different matches, they will have a different groupId each time.
 - matchId - ID to identify match. There are no matches that are in both the training and testing set.
 
**To simplify the algorithm, we decide to remove all statistics related to ELO ranking**

 - killPoints - Kills-based external ranking of player. (Think of this as an Elo ranking where only kills matter.) If there is a value other than -1 in rankPoints, then any 0 in killPoints should be treated as a “None”.
 - rankPoints - Elo-like ranking of player. This ranking is inconsistent and is being deprecated in the API’s next version, so use with caution. Value of -1 takes place of “None”.
 - winPoints - Win-based external ranking of player. (Think of this as an Elo ranking where only winning matters.) If there is a value other than -1 in rankPoints, then any 0 in winPoints should be treated as a “None”.
 
**Since there are some statistics may confuse the results due to complex match cases, we remove them instead of dealing with them**
 - longestKill - Longest distance between player and player killed at time of death. This may be misleading, as downing a player and driving away may lead to a large longestKill stat.
 - matchType - String identifying the game mode that the data comes from. The standard modes are “solo”, “duo”, “squad”, “solo-fpp”, “duo-fpp”, and “squad-fpp”; other modes are from events or custom matches.

In [0]:
trainDF = originDF.drop('Id', 'groupId', 'matchId', 'killPoints', 'rankPoints', 'winPoints', 'longestKill', 'matchType')
display(trainDF)

assists boosts damageDealt DBNOs headshotKills heals killPlace kills killStreaks matchDuration maxPlace numGroups revives rideDistance roadKills swimDistance teamKills vehicleDestroys walkDistance weaponsAcquired winPlacePerc 0.0 0.0 0.0 0.0 0.0 0.0 60.0 0.0 0.0 1306.0 28.0 26.0 0.0 0.0 0.0 0.0 0.0 0.0 244.8 1.0 0.4444 0.0 0.0 91.47 0.0 0.0 0.0 57.0 0.0 0.0 1777.0 26.0 25.0 0.0 0.0045 0.0 11.04 0.0 0.0 1434.0 5.0 0.64 1.0 0.0 68.0 0.0 0.0 0.0 47.0 0.0 0.0 1318.0 50.0 47.0 0.0 0.0 0.0 0.0 0.0 0.0 161.8 2.0 0.7755 0.0 0.0 32.9 0.0 0.0 0.0 75.0 0.0 0.0 1436.0 31.0 30.0 0.0 0.0 0.0 0.0 0.0 0.0 202.7 3.0 0.1667 0.0 0.0 100.0 0.0 0.0 0.0 45.0 1.0 1.0 1424.0 97.0 95.0 0.0 0.0 0.0 0.0 0.0 0.0 49.75 2.0 0.1875 0.0 0.0 100.0 1.0 1.0 0.0 44.0 1.0 1.0 1395.0 28.0 28.0 0.0 0.0 0.0 0.0 0.0 0.0 34.7 1.0 0.037 0.0 0.0 0.0 0.0 0.0 0.0 96.0 0.0 0.0 1316.0 28.0 28.0 0.0 0.0 0.0 0.0 0.0 0.0 13.5 1.0 0.0 0.0 0.0 8.538 0.0 0.0 0.0 48.0 0.0 0.0 1967.0 96.0 92.0 0.0 2004.0 0.0 0.0 0.0 0.0 1089.0 6.0 0.7368 0.0 0.0 51.6 0.0 0.0 0.0 64.0 0.0 0.0 1375.0 28.0 27.0 0.0 0.0 0.0 0.0 0.0 0.0 799.9 4.0 0.3704 0.0 0.0 37.27 0.0 0.0 0.0 74.0 0.0 0.0 1930.0 29.0 27.0 0.0 0.0 0.0 0.0 0.0 0.0 65.67 1.0 0.2143 0.0 0.0 28.38 0.0 0.0 0.0 75.0 0.0 0.0 1811.0 29.0 29.0 0.0 0.0 0.0 0.0 0.0 0.0 868.3 9.0 0.3929 0.0 0.0 137.9 1.0 0.0 0.0 64.0 0.0 0.0 1384.0 48.0 46.0 0.0 0.0 0.0 0.0 0.0 0.0 451.7 1.0 0.4043 0.0 0.0 0.0 0.0 0.0 0.0 37.0 0.0 0.0 1774.0 29.0 28.0 0.0 6639.0 0.0 0.0 0.0 0.0 2784.0 6.0 0.9286 0.0 1.0 324.2 0.0 1.0 5.0 5.0 4.0 1.0 1886.0 97.0 94.0 0.0 1228.0 0.0 76.84 0.0 0.0 2050.0 6.0 0.875 0.0 1.0 122.8 1.0 0.0 2.0 25.0 1.0 1.0 1458.0 31.0 30.0 1.0 1237.0 0.0 60.29 0.0 0.0 1666.0 5.0 0.9 0.0 0.0 80.71 1.0 0.0 0.0 72.0 0.0 0.0 1313.0 48.0 48.0 1.0 0.0 0.0 0.0 0.0 0.0 105.1 5.0 0.2766 0.0 2.0 81.71 1.0 0.0 14.0 25.0 1.0 1.0 1882.0 27.0 26.0 0.0 519.9 0.0 0.0 0.0 0.0 3674.0 7.0 0.7308 0.0 3.0 254.3 0.0 0.0 12.0 13.0 2.0 1.0 1371.0 96.0 95.0 0.0 2367.0 0.0 15.29 0.0 0.0 1787.0 3.0 0.8211 0.0 0.0 0.0 0.0 0.0 0.0 79.0 0.0 0.0 1301.0 27.0 26.0 0.0 0.0 0.0 0.0 0.0 0.0 137.4 2.0 0.1923 0.0 1.0 65.28 0.0 0.0 1.0 48.0 0.0 0.0 1322.0 30.0 28.0 0.0 0.0 0.0 20.85 0.0 0.0 3310.0 3.0 0.931 0.0 4.0 269.1 0.0 1.0 8.0 18.0 2.0 1.0 1890.0 48.0 48.0 1.0 2734.0 0.0 0.0 0.0 0.0 1794.0 5.0 0.6383 0.0 0.0 158.7 1.0 0.0 0.0 75.0 0.0 0.0 1853.0 29.0 28.0 0.0 0.0 0.0 0.0 0.0 0.0 580.1 2.0 0.2143 0.0 1.0 192.3 1.0 2.0 3.0 15.0 2.0 1.0 1775.0 49.0 48.0 0.0 2332.0 0.0 0.0 0.0 0.0 1264.0 4.0 0.75 0.0 6.0 1011.0 6.0 2.0 2.0 2.0 9.0 2.0 1857.0 50.0 45.0 0.0 4860.0 0.0 0.0 0.0 0.0 2727.0 7.0 0.9592 0.0 3.0 327.6 4.0 1.0 1.0 3.0 4.0 2.0 1454.0 27.0 27.0 0.0 0.0 0.0 0.0 0.0 0.0 3503.0 4.0 0.9231 1.0 4.0 558.6 3.0 0.0 4.0 11.0 3.0 2.0 1841.0 47.0 45.0 0.0 1183.0 0.0 0.0 0.0 0.0 2711.0 7.0 0.8696 0.0 0.0 44.28 0.0 0.0 0.0 78.0 0.0 0.0 1879.0 27.0 25.0 0.0 0.0 0.0 0.0 0.0 0.0 15.3 0.0 0.1154 0.0 4.0 381.2 2.0 1.0 2.0 7.0 4.0 2.0 1896.0 48.0 46.0 0.0 1798.0 0.0 0.0 0.0 0.0 1933.0 4.0 0.7234 0.0 6.0 345.6 2.0 1.0 1.0 6.0 4.0 1.0 1339.0 28.0 28.0 0.0 0.0 0.0 0.0 0.0 0.0 3855.0 4.0 0.963 0.0 0.0 0.0 0.0 0.0 0.0 87.0 0.0 0.0 1530.0 46.0 44.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 36.73 0.0 0.0 0.0 62.0 0.0 0.0 1901.0 48.0 45.0 0.0 0.0 0.0 0.0 0.0 0.0 455.5 5.0 0.4255 0.0 0.0 0.0 0.0 0.0 0.0 80.0 0.0 0.0 1371.0 31.0 31.0 0.0 0.0 0.0 0.0 0.0 0.0 93.11 2.0 0.0667 0.0 3.0 0.0 0.0 0.0 1.0 61.0 0.0 0.0 1913.0 49.0 46.0 0.0 1154.0 0.0 0.0 1.0 0.0 1240.0 7.0 0.5 0.0 2.0 65.47 0.0 0.0 2.0 61.0 0.0 0.0 1348.0 26.0 26.0 3.0 0.0 0.0 0.0 0.0 0.0 751.8 8.0 0.52 0.0 1.0 439.9 4.0 0.0 1.0 34.0 1.0 1.0 1905.0 28.0 26.0 0.0 0.0 0.0 0.0 0.0 0.0 353.5 2.0 0.2222 0.0 0.0 136.9 0.0 0.0 0.0 37.0 1.0 1.0 1425.0 96.0 94.0 0.0 0.0 0.0 0.0 0.0 0.0 270.7 1.0 0.3474 0.0 1.0 100.0 0.0 0.0 0.0 24.0 1.0 1.0 1398.0 92.0 89.0 0.0 0.0 0.0 61.0 0.0 0.0 1528.0 3.0 0.8462 0.0 0.0 0.0 0.0 0.0 0.0 82.0 0.0 0.0 1518.0 46.0 44.0 0.0 0.0 0.0 0.0 0.0 0.0 60.93 1.0 0.0889 1.0 0.0 318.0 2.0 0.0 0.0 5.0 4.0 2.0 1365.0 48.0 45.0 0.0 0.0 0.0 0.0

**Then we register our DataFrame as an SQL table named "project.dataset", and do some basic statistical analyses of all the columns.**

In [0]:
sqlContext.sql("DROP TABLE IF EXISTS project_dataset")
dbutils.fs.rm("dbfs:/user/hive/warehouse/project_dataset", True)
sqlContext.registerDataFrameAsTable(trainDF, "project_dataset")

**Now we need to examine some of statistics that are unrelated to the winplace percentage.**
**We choose walkDistance, killPlace, matchDuration, maxPlace, numGroups, teamKills as our example.**

In [0]:
%sql
select walkDistance, winPlacePerc from project_dataset

walkDistance winPlacePerc 244.8 0.4444 1434.0 0.64 161.8 0.7755 202.7 0.1667 49.75 0.1875 34.7 0.037 13.5 0.0 1089.0 0.7368 799.9 0.3704 65.67 0.2143 868.3 0.3929 451.7 0.4043 2784.0 0.9286 2050.0 0.875 1666.0 0.9 105.1 0.2766 3674.0 0.7308 1787.0 0.8211 137.4 0.1923 3310.0 0.931 1794.0 0.6383 580.1 0.2143 1264.0 0.75 2727.0 0.9592 3503.0 0.9231 2711.0 0.8696 15.3 0.1154 1933.0 0.7234 3855.0 0.963 0.0 0.0 455.5 0.4255 93.11 0.0667 1240.0 0.5 751.8 0.52 353.5 0.2222 270.7 0.3474 1528.0 0.8462 60.93 0.0889 2303.0 0.7021 133.4 0.1481 327.3 0.32 128.8 0.0217 52.52 0.0357 534.1 0.3958 2576.0 0.7872 2003.0 0.8077 68.56 0.1786 79.33 0.125 248.6 0.2947 991.7 0.3571 2037.0 0.5 577.2 0.3793 212.9 0.25 143.8 0.2174 3663.0 0.8214 2190.0 0.64 3078.0 0.8929 687.7 0.5517 127.7 0.1111 1606.0 0.6809 3182.0 0.72 2593.0 0.9091 9.907 0.0 471.9 0.2245 735.0 0.4468 1378.0 0.5417 4286.0 0.963 3120.0 1.0 194.1 0.1875 231.7 0.1573 3580.0 0.617 2135.0 0.7667 3165.0 0.9655 4321.0 0.8571 27.06 0.0 1488.0 0.6522 576.7 0.2857 1749.0 0.7609 140.1 0.0769 589.3 0.25 103.0 0.0714 3563.0 0.9796 982.5 0.2963 1150.0 0.5 638.0 0.5714 3683.0 0.8958 14.85 0.2414 2666.0 0.7143 827.6 0.3462 50.82 0.1489 547.5 0.2692 292.6 0.1075 258.8 0.2292 747.9 0.5 302.3 0.3958 1089.0 0.25 565.5 0.4231 3309.0 0.7586 822.1 0.5106 249.7 0.1071 1913.0 0.9355 1766.0 0.9891 67.54 0.1429 68.9 0.0 838.7 0.7021 1704.0 0.6923 1542.0 0.883 2708.0 1.0 615.0 0.4643 20.34 0.0 43.67 0.0816 855.1 0.6526 643.9 0.3214 2693.0 0.871 110.7 0.0833 943.3 0.2308 0.0 0.0 148.2 0.4105 2635.0 1.0 304.5 0.3409 1493.0 0.6327 870.9 0.3878 1118.0 0.6667 2814.0 1.0 1677.0 0.5926 565.9 0.3462 536.9 0.6667 1104.0 0.5625 151.4 0.037 2532.0 0.68 3514.0 0.8148 1575.0 0.9643 92.3 0.4583 79.86 0.0385 1497.0 0.6327 145.6 0.069 195.8 0.3542 101.8 0.3474 33.85 0.2083 23.19 0.0825 184.8 0.1786 1673.0 0.7604 2363.0 0.9255 165.5 0.0 129.8 0.1071 672.6 0.3846 133.9 0.2727 368.2 0.4421 62.88 0.037 5.722 0.1429 22.6 0.1071 0.0 0.0 1313.0 0.6087 93.32 0.125 2907.0 0.7333 462.0 0.3542 1519.0 0.4074 3592.0 0.9222 1675.0 0.6538 8.291 0.1957 203.9 0.0357 684.4 0.5208 331.6 0.2292 30.99 0.2857 420.8 0.8889 3171.0 0.9592 93.57 0.0769 2813.0 0.9259 208.4 0.037 1728.0 0.6774 1042.0 0.5833 798.1 0.4222 3875.0 0.9787 100.6 0.1786 2385.0 0.8667 179.0 0.5 1984.0 0.8085 101.9 0.1579 3861.0 0.7333 61.45 0.0357 2054.0 0.6296 3799.0 0.8966 2487.0 0.9375 260.9 0.1852 3875.0 0.84 110.0 0.1034 2422.0 1.0 20.9 0.0 942.8 0.1481 471.7 0.4138 145.0 0.04 296.3 0.8519 44.6 0.3448 664.9 0.4444 848.7 0.6082 1012.0 0.9263 2557.0 0.6591 1200.0 0.5385 581.3 0.4444 3205.0 0.9259 84.58 0.0741 277.8 0.4375 8.214 0.1313 10.16 0.0 155.9 0.037 1287.0 0.6667 189.7 0.2128 322.0 0.3462 76.6 0.1111 6.6 0.1146 1313.0 0.5385 656.2 0.5625 787.0 0.5 68.94 0.08 82.96 0.2963 1072.0 0.6071 100.6 0.1111 350.8 0.3333 515.6 0.3125 324.5 0.2083 3305.0 0.963 2247.0 0.5385 9.003 0.24 352.4 0.1071 550.9 0.2857 2240.0 0.7143 122.0 0.0769 842.3 0.413 339.7 0.2234 2697.0 1.0 3440.0 0.6889 201.3 0.122 257.0 0.1702 3409.0 0.92 250.9 0.1379 23.23 0.0435 270.7 0.1111 0.0 0.0 89.92 0.2292 240.1 0.1538 2500.0 0.7692 223.9 0.1702 26.54 0.1667 2534.0 0.8925 2065.0 0.6923 1327.0 0.5185 520.9 0.1034 219.4 0.3404 136.1 0.1852 3264.0 0.9091 85.47 0.25 960.3 0.4828 499.6 0.4444 305.8 0.2 308.0 0.3704 73.7 0.4255 2631.0 0.7692 99.42 0.1429 1844.0 0.8667 2477.0 0.8889 2484.0 0.8276 1468.0 0.6739 3157.0 0.9231 168.1 0.383 393.2 0.3448 3187.0 0.8 2237.0 0.9574 780.2 0.6429 876.0 0.5185 931.4 0.4815 2401.0 0.7667 437.8 0.3704 2807.0 0.8667 1374.0 0.3462 2346.0 0.8687 50.01 0.1224 599.2 0.5745 1019.0 0.4815 1403.0 0.6154 79.89 0.0714 1831.0 0.8148 538.4 0.44 160.4 0.0606 0.0 0.1277 3039.0 0.9 2993.0 0.8148 3293.0 0.9588 3451.0 0.7708 1905.0 0.5714 1865.0 0.5769 683.5 0.3333 675.0 0.4205 94.53 0.0385 1949.0 0.9643 1317.0 0.8043 854.4 0.5714 13.4 0.0652 1846.0 0.7857 2633.0 0.9896 54.12 0.0833 3503.0 1.0 93.56 0.0741 267.9 0.5263 7.813 0.0417 1632.0 0.9259 3395.0 0.9615 420.1 0.2963 1772.0 0.3462 167.7 0.

In [0]:
%sql
select killPlace, winPlacePerc from project_dataset

killPlace winPlacePerc 60.0 0.4444 57.0 0.64 47.0 0.7755 75.0 0.1667 45.0 0.1875 44.0 0.037 96.0 0.0 48.0 0.7368 64.0 0.3704 74.0 0.2143 75.0 0.3929 64.0 0.4043 37.0 0.9286 5.0 0.875 25.0 0.9 72.0 0.2766 25.0 0.7308 13.0 0.8211 79.0 0.1923 48.0 0.931 18.0 0.6383 75.0 0.2143 15.0 0.75 2.0 0.9592 3.0 0.9231 11.0 0.8696 78.0 0.1154 7.0 0.7234 6.0 0.963 87.0 0.0 62.0 0.4255 80.0 0.0667 61.0 0.5 61.0 0.52 34.0 0.2222 37.0 0.3474 24.0 0.8462 82.0 0.0889 5.0 0.7021 73.0 0.1481 31.0 0.32 86.0 0.0217 46.0 0.0357 12.0 0.3958 27.0 0.7872 13.0 0.8077 77.0 0.1786 79.0 0.125 19.0 0.2947 10.0 0.3571 63.0 0.5 67.0 0.3793 36.0 0.25 37.0 0.2174 47.0 0.8214 4.0 0.64 6.0 0.8929 29.0 0.5517 47.0 0.1111 29.0 0.6809 16.0 0.72 8.0 0.9091 41.0 0.0 79.0 0.2245 21.0 0.4468 27.0 0.5417 2.0 0.963 4.0 1.0 77.0 0.1875 38.0 0.1573 55.0 0.617 5.0 0.7667 49.0 0.9655 12.0 0.8571 91.0 0.0 54.0 0.6522 40.0 0.2857 29.0 0.7609 86.0 0.0769 69.0 0.25 92.0 0.0714 23.0 0.9796 71.0 0.2963 30.0 0.5 27.0 0.5714 27.0 0.8958 47.0 0.2414 10.0 0.7143 20.0 0.3462 81.0 0.1489 56.0 0.2692 84.0 0.1075 40.0 0.2292 66.0 0.5 64.0 0.3958 52.0 0.25 62.0 0.4231 49.0 0.7586 2.0 0.5106 85.0 0.1071 7.0 0.9355 4.0 0.9891 84.0 0.1429 94.0 0.0 40.0 0.7021 49.0 0.6923 50.0 0.883 38.0 1.0 41.0 0.4643 96.0 0.0 83.0 0.0816 58.0 0.6526 72.0 0.3214 11.0 0.871 75.0 0.0833 16.0 0.2308 68.0 0.0 65.0 0.4105 45.0 1.0 60.0 0.3409 54.0 0.6327 65.0 0.3878 31.0 0.6667 4.0 1.0 30.0 0.5926 21.0 0.3462 15.0 0.6667 28.0 0.5625 82.0 0.037 11.0 0.68 26.0 0.8148 6.0 0.9643 61.0 0.4583 87.0 0.0385 51.0 0.6327 83.0 0.069 35.0 0.3542 68.0 0.3474 25.0 0.2083 90.0 0.0825 41.0 0.1786 49.0 0.7604 12.0 0.9255 91.0 0.0 89.0 0.1071 12.0 0.3846 65.0 0.2727 65.0 0.4421 44.0 0.037 83.0 0.1429 84.0 0.1071 92.0 0.0 55.0 0.6087 81.0 0.125 44.0 0.7333 42.0 0.3542 68.0 0.4074 12.0 0.9222 30.0 0.6538 77.0 0.1957 87.0 0.0357 64.0 0.5208 77.0 0.2292 71.0 0.2857 40.0 0.8889 46.0 0.9592 85.0 0.0769 44.0 0.9259 85.0 0.037 5.0 0.6774 59.0 0.5833 62.0 0.4222 18.0 0.9787 82.0 0.1786 25.0 0.8667 34.0 0.5 29.0 0.8085 83.0 0.1579 53.0 0.7333 84.0 0.0357 33.0 0.6296 9.0 0.8966 5.0 0.9375 83.0 0.1852 7.0 0.84 88.0 0.1034 2.0 1.0 87.0 0.0 83.0 0.1481 35.0 0.4138 81.0 0.04 8.0 0.8519 74.0 0.3448 30.0 0.4444 31.0 0.6082 23.0 0.9263 18.0 0.6591 11.0 0.5385 75.0 0.4444 4.0 0.9259 88.0 0.0741 67.0 0.4375 88.0 0.1313 43.0 0.0 92.0 0.037 24.0 0.6667 80.0 0.2128 70.0 0.3462 68.0 0.1111 86.0 0.1146 28.0 0.5385 62.0 0.5625 17.0 0.5 43.0 0.08 80.0 0.2963 3.0 0.6071 77.0 0.1111 41.0 0.3333 41.0 0.3125 66.0 0.2083 8.0 0.963 56.0 0.5385 77.0 0.24 79.0 0.1071 69.0 0.2857 9.0 0.7143 79.0 0.0769 17.0 0.413 75.0 0.2234 4.0 1.0 53.0 0.6889 73.0 0.122 80.0 0.1702 46.0 0.92 77.0 0.1379 83.0 0.0435 89.0 0.1111 94.0 0.0 79.0 0.2292 78.0 0.1538 23.0 0.7692 82.0 0.1702 79.0 0.1667 21.0 0.8925 30.0 0.6923 33.0 0.5185 81.0 0.1034 68.0 0.3404 75.0 0.1852 10.0 0.9091 35.0 0.25 35.0 0.4828 28.0 0.4444 76.0 0.2 19.0 0.3704 60.0 0.4255 48.0 0.7692 84.0 0.1429 45.0 0.8667 45.0 0.8889 2.0 0.8276 10.0 0.6739 2.0 0.9231 68.0 0.383 21.0 0.3448 16.0 0.8 24.0 0.9574 29.0 0.6429 59.0 0.5185 62.0 0.4815 47.0 0.7667 20.0 0.3704 44.0 0.8667 21.0 0.3462 45.0 0.8687 84.0 0.1224 28.0 0.5745 18.0 0.4815 46.0 0.6154 92.0 0.0714 2.0 0.8148 21.0 0.44 88.0 0.0606 84.0 0.1277 25.0 0.9 43.0 0.8148 23.0 0.9588 54.0 0.7708 53.0 0.5714 59.0 0.5769 16.0 0.3333 11.0 0.4205 89.0 0.0385 36.0 0.9643 5.0 0.8043 18.0 0.5714 89.0 0.0652 24.0 0.7857 7.0 0.9896 88.0 0.0833 1.0 1.0 90.0 0.0741 58.0 0.5263 87.0 0.0417 21.0 0.9259 13.0 0.9615 76.0 0.2963 34.0 0.3462 44.0 0.2653 55.0 0.48 80.0 0.037 29.0 0.5926 74.0 0.2917 8.0 0.7931 76.0 0.2857 70.0 0.3043 76.0 0.1562 57.0 0.4516 17.0 0.3103 80.0 0.2414 8.0 0.1379 9.0 0.7586 87.0 0.0851 36.0 0.1786 66.0 0.4227 91.0 0.0385 82.0 0.1875 16.0 0.6458 79.0 0.2128 27.0 0.7667 2.0 0.7447 2.0 0.8889 9.0 0.8298 21.0 0.5208 19.0 0.4947 35.0 0.6042 30.0 0.7551 16.0 0.8298 72.0 0.1852 11.0 0.766 28.0 0.5333 29.0 1.0 31.0 0.4167 82.0 0.2083 11.0 0.8958 55.0 0.4615 24.0 0.931 2

In [0]:
%sql
select matchDuration, winPlacePerc from project_dataset

matchDuration winPlacePerc 1306.0 0.4444 1777.0 0.64 1318.0 0.7755 1436.0 0.1667 1424.0 0.1875 1395.0 0.037 1316.0 0.0 1967.0 0.7368 1375.0 0.3704 1930.0 0.2143 1811.0 0.3929 1384.0 0.4043 1774.0 0.9286 1886.0 0.875 1458.0 0.9 1313.0 0.2766 1882.0 0.7308 1371.0 0.8211 1301.0 0.1923 1322.0 0.931 1890.0 0.6383 1853.0 0.2143 1775.0 0.75 1857.0 0.9592 1454.0 0.9231 1841.0 0.8696 1879.0 0.1154 1896.0 0.7234 1339.0 0.963 1530.0 0.0 1901.0 0.4255 1371.0 0.0667 1913.0 0.5 1348.0 0.52 1905.0 0.2222 1425.0 0.3474 1398.0 0.8462 1518.0 0.0889 1365.0 0.7021 1222.0 0.1481 1303.0 0.32 1843.0 0.0217 1813.0 0.0357 1367.0 0.3958 1905.0 0.7872 1449.0 0.8077 1352.0 0.1786 1391.0 0.125 2067.0 0.2947 1909.0 0.3571 1815.0 0.5 1355.0 0.3793 1767.0 0.25 1343.0 0.2174 1867.0 0.8214 1374.0 0.64 1423.0 0.8929 1357.0 0.5517 1196.0 0.1111 1959.0 0.6809 1327.0 0.72 1289.0 0.9091 1278.0 0.0 1945.0 0.2245 1388.0 0.4468 1739.0 0.5417 1308.0 0.963 1297.0 1.0 1192.0 0.1875 2042.0 0.1573 2048.0 0.617 1404.0 0.7667 1358.0 0.9655 1827.0 0.8571 1376.0 0.0 1867.0 0.6522 1350.0 0.2857 1892.0 0.7609 1372.0 0.0769 2007.0 0.25 1817.0 0.0714 1972.0 0.9796 1809.0 0.2963 1318.0 0.5 1442.0 0.5714 1777.0 0.8958 1341.0 0.2414 1665.0 0.7143 1892.0 0.3462 1493.0 0.1489 1719.0 0.2692 1999.0 0.1075 1423.0 0.2292 1391.0 0.5 1399.0 0.3958 1796.0 0.25 1913.0 0.4231 1287.0 0.7586 1380.0 0.5106 1743.0 0.1071 1423.0 0.9355 1465.0 0.9891 1387.0 0.1429 1360.0 0.0 1321.0 0.7021 1403.0 0.6923 1441.0 0.883 1708.0 1.0 1369.0 0.4643 1923.0 0.0 1399.0 0.0816 1356.0 0.6526 1309.0 0.3214 1840.0 0.871 1885.0 0.0833 1822.0 0.2308 1414.0 0.0 1290.0 0.4105 1388.0 1.0 1878.0 0.3409 2048.0 0.6327 1471.0 0.3878 1871.0 0.6667 1405.0 1.0 1370.0 0.5926 1283.0 0.3462 1349.0 0.6667 1448.0 0.5625 1340.0 0.037 1905.0 0.68 1882.0 0.8148 1910.0 0.9643 1394.0 0.4583 1357.0 0.0385 1769.0 0.6327 1334.0 0.069 1381.0 0.3542 1349.0 0.3474 1906.0 0.2083 1453.0 0.0825 1400.0 0.1786 1935.0 0.7604 1435.0 0.9255 1307.0 0.0 1374.0 0.1071 1862.0 0.3846 1516.0 0.2727 1384.0 0.4421 1375.0 0.037 1278.0 0.1429 1720.0 0.1071 1377.0 0.0 1335.0 0.6087 1327.0 0.125 1895.0 0.7333 1354.0 0.3542 1879.0 0.4074 1801.0 0.9222 1898.0 0.6538 1950.0 0.1957 1839.0 0.0357 1262.0 0.5208 1368.0 0.2292 1808.0 0.2857 1380.0 0.8889 1391.0 0.9592 1402.0 0.0769 1703.0 0.9259 1785.0 0.037 1339.0 0.6774 1959.0 0.5833 1970.0 0.4222 1420.0 0.9787 1351.0 0.1786 1380.0 0.8667 1413.0 0.5 1377.0 0.8085 1889.0 0.1579 1883.0 0.7333 1302.0 0.0357 1386.0 0.6296 1780.0 0.8966 1882.0 0.9375 1805.0 0.1852 1824.0 0.84 1344.0 0.1034 1387.0 1.0 1353.0 0.0 1756.0 0.1481 1108.0 0.4138 1319.0 0.04 1851.0 0.8519 1359.0 0.3448 1915.0 0.4444 1370.0 0.6082 1869.0 0.9263 1888.0 0.6591 1758.0 0.5385 1393.0 0.4444 1270.0 0.9259 1331.0 0.0741 1333.0 0.4375 1369.0 0.1313 1470.0 0.0 1301.0 0.037 1765.0 0.6667 1318.0 0.2128 1397.0 0.3462 1821.0 0.1111 1394.0 0.1146 1318.0 0.5385 1322.0 0.5625 1959.0 0.5 1357.0 0.08 1326.0 0.2963 1301.0 0.6071 1517.0 0.1111 1897.0 0.3333 1376.0 0.3125 1303.0 0.2083 1825.0 0.963 1512.0 0.5385 1532.0 0.24 1397.0 0.1071 1835.0 0.2857 1307.0 0.7143 1435.0 0.0769 1776.0 0.413 1963.0 0.2234 1394.0 1.0 1361.0 0.6889 1486.0 0.122 1413.0 0.1702 1396.0 0.92 1321.0 0.1379 1809.0 0.0435 1396.0 0.1111 1358.0 0.0 1406.0 0.2292 1420.0 0.1538 1338.0 0.7692 1822.0 0.1702 1360.0 0.1667 1462.0 0.8925 1829.0 0.6923 1273.0 0.5185 1825.0 0.1034 1448.0 0.3404 1416.0 0.1852 1325.0 0.9091 1461.0 0.25 1383.0 0.4828 1369.0 0.4444 1898.0 0.2 1415.0 0.3704 1403.0 0.4255 1380.0 0.7692 1351.0 0.1429 1314.0 0.8667 1383.0 0.8889 1336.0 0.8276 1968.0 0.6739 1322.0 0.9231 1404.0 0.383 1787.0 0.3448 1950.0 0.8 1325.0 0.9574 1902.0 0.6429 1378.0 0.5185 1384.0 0.4815 1206.0 0.7667 1227.0 0.3704 1268.0 0.8667 1894.0 0.3462 1809.0 0.8687 1960.0 0.1224 1378.0 0.5745 1384.0 0.4815 1799.0 0.6154 1796.0 0.0714 1806.0 0.8148 1341.0 0.44 1913.0 0.0606 1797.0 0.1277 1889.0 0.9 1879.0 0.8148 1381.0 0.9588 1805.0 0.7708 1369.0 0.5714 1369.0 0.5769 1406.0 0.3333 1896.0 0.4205 1391.0 0.0385 1846.0 0

In [0]:
%sql
select maxPlace, winPlacePerc from project_dataset

maxPlace winPlacePerc 28.0 0.4444 26.0 0.64 50.0 0.7755 31.0 0.1667 97.0 0.1875 28.0 0.037 28.0 0.0 96.0 0.7368 28.0 0.3704 29.0 0.2143 29.0 0.3929 48.0 0.4043 29.0 0.9286 97.0 0.875 31.0 0.9 48.0 0.2766 27.0 0.7308 96.0 0.8211 27.0 0.1923 30.0 0.931 48.0 0.6383 29.0 0.2143 49.0 0.75 50.0 0.9592 27.0 0.9231 47.0 0.8696 27.0 0.1154 48.0 0.7234 28.0 0.963 46.0 0.0 48.0 0.4255 31.0 0.0667 49.0 0.5 26.0 0.52 28.0 0.2222 96.0 0.3474 92.0 0.8462 46.0 0.0889 48.0 0.7021 28.0 0.1481 26.0 0.32 47.0 0.0217 29.0 0.0357 49.0 0.3958 48.0 0.7872 27.0 0.8077 29.0 0.1786 49.0 0.125 96.0 0.2947 29.0 0.3571 31.0 0.5 30.0 0.3793 29.0 0.25 47.0 0.2174 29.0 0.8214 26.0 0.64 29.0 0.8929 30.0 0.5517 28.0 0.1111 95.0 0.6809 26.0 0.72 45.0 0.9091 27.0 0.0 99.0 0.2245 48.0 0.4468 25.0 0.5417 28.0 0.963 29.0 1.0 49.0 0.1875 90.0 0.1573 48.0 0.617 31.0 0.7667 30.0 0.9655 29.0 0.8571 27.0 0.0 47.0 0.6522 29.0 0.2857 47.0 0.7609 27.0 0.0769 29.0 0.25 29.0 0.0714 50.0 0.9796 28.0 0.2963 29.0 0.5 50.0 0.5714 49.0 0.8958 30.0 0.2414 29.0 0.7143 27.0 0.3462 95.0 0.1489 27.0 0.2692 94.0 0.1075 49.0 0.2292 49.0 0.5 49.0 0.3958 21.0 0.25 27.0 0.4231 30.0 0.7586 48.0 0.5106 29.0 0.1071 94.0 0.9355 93.0 0.9891 29.0 0.1429 27.0 0.0 48.0 0.7021 27.0 0.6923 95.0 0.883 26.0 1.0 29.0 0.4643 96.0 0.0 50.0 0.0816 96.0 0.6526 29.0 0.3214 32.0 0.871 49.0 0.0833 27.0 0.2308 41.0 0.0 96.0 0.4105 49.0 1.0 45.0 0.3409 50.0 0.6327 99.0 0.3878 49.0 0.6667 29.0 1.0 28.0 0.5926 27.0 0.3462 94.0 0.6667 49.0 0.5625 28.0 0.037 26.0 0.68 28.0 0.8148 29.0 0.9643 49.0 0.4583 27.0 0.0385 50.0 0.6327 30.0 0.069 49.0 0.3542 96.0 0.3474 49.0 0.2083 98.0 0.0825 29.0 0.1786 97.0 0.7604 95.0 0.9255 28.0 0.0 29.0 0.1071 27.0 0.3846 34.0 0.2727 96.0 0.4421 28.0 0.037 29.0 0.1429 29.0 0.1071 48.0 0.0 47.0 0.6087 49.0 0.125 46.0 0.7333 49.0 0.3542 28.0 0.4074 91.0 0.9222 27.0 0.6538 47.0 0.1957 29.0 0.0357 49.0 0.5208 49.0 0.2292 29.0 0.2857 28.0 0.8889 50.0 0.9592 27.0 0.0769 28.0 0.9259 28.0 0.037 32.0 0.6774 49.0 0.5833 46.0 0.4222 48.0 0.9787 29.0 0.1786 31.0 0.8667 27.0 0.5 48.0 0.8085 96.0 0.1579 31.0 0.7333 29.0 0.0357 28.0 0.6296 30.0 0.8966 97.0 0.9375 28.0 0.1852 26.0 0.84 30.0 0.1034 98.0 1.0 44.0 0.0 28.0 0.1481 30.0 0.4138 26.0 0.04 28.0 0.8519 30.0 0.3448 91.0 0.4444 98.0 0.6082 96.0 0.9263 45.0 0.6591 27.0 0.5385 28.0 0.4444 28.0 0.9259 28.0 0.0741 49.0 0.4375 100.0 0.1313 46.0 0.0 28.0 0.037 28.0 0.6667 48.0 0.2128 27.0 0.3462 46.0 0.1111 97.0 0.1146 27.0 0.5385 97.0 0.5625 25.0 0.5 26.0 0.08 28.0 0.2963 29.0 0.6071 28.0 0.1111 46.0 0.3333 49.0 0.3125 25.0 0.2083 28.0 0.963 27.0 0.5385 26.0 0.24 29.0 0.1071 50.0 0.2857 29.0 0.7143 27.0 0.0769 47.0 0.413 95.0 0.2234 29.0 1.0 46.0 0.6889 42.0 0.122 48.0 0.1702 26.0 0.92 30.0 0.1379 24.0 0.0435 28.0 0.1111 29.0 0.0 49.0 0.2292 27.0 0.1538 27.0 0.7692 48.0 0.1702 25.0 0.1667 94.0 0.8925 27.0 0.6923 28.0 0.5185 30.0 0.1034 48.0 0.3404 28.0 0.1852 45.0 0.9091 49.0 0.25 30.0 0.4828 28.0 0.4444 46.0 0.2 28.0 0.3704 48.0 0.4255 27.0 0.7692 29.0 0.1429 31.0 0.8667 46.0 0.8889 30.0 0.8276 47.0 0.6739 27.0 0.9231 48.0 0.383 30.0 0.3448 16.0 0.8 48.0 0.9574 29.0 0.6429 28.0 0.5185 28.0 0.4815 31.0 0.7667 28.0 0.3704 31.0 0.8667 27.0 0.3462 100.0 0.8687 50.0 0.1224 48.0 0.5745 28.0 0.4815 27.0 0.6154 29.0 0.0714 28.0 0.8148 26.0 0.44 34.0 0.0606 48.0 0.1277 31.0 0.9 28.0 0.8148 98.0 0.9588 49.0 0.7708 29.0 0.5714 27.0 0.5769 28.0 0.3333 89.0 0.4205 27.0 0.0385 29.0 0.9643 93.0 0.8043 29.0 0.5714 47.0 0.0652 29.0 0.7857 97.0 0.9896 49.0 0.0833 31.0 1.0 28.0 0.0741 96.0 0.5263 25.0 0.0417 28.0 0.9259 27.0 0.9615 28.0 0.2963 27.0 0.3462 99.0 0.2653 26.0 0.48 28.0 0.037 28.0 0.5926 49.0 0.2917 30.0 0.7931 50.0 0.2857 47.0 0.3043 33.0 0.1562 32.0 0.4516 30.0 0.3103 30.0 0.2414 30.0 0.1379 30.0 0.7586 48.0 0.0851 29.0 0.1786 98.0 0.4227 27.0 0.0385 49.0 0.1875 49.0 0.6458 48.0 0.2128 31.0 0.7667 48.0 0.7447 28.0 0.8889 48.0 0.8298 97.0 0.5208 96.0 0.4947 49.0 0.6042 50.0 0.7551 48.0 0.8298 28.0 0.1852 48.0 0.766 46.0 0.5333 46.0 1.0 25.0 0.4167 49.0 0.

In [0]:
%sql
select numGroups, winPlacePerc from project_dataset

numGroups winPlacePerc 26.0 0.4444 25.0 0.64 47.0 0.7755 30.0 0.1667 95.0 0.1875 28.0 0.037 28.0 0.0 92.0 0.7368 27.0 0.3704 27.0 0.2143 29.0 0.3929 46.0 0.4043 28.0 0.9286 94.0 0.875 30.0 0.9 48.0 0.2766 26.0 0.7308 95.0 0.8211 26.0 0.1923 28.0 0.931 48.0 0.6383 28.0 0.2143 48.0 0.75 45.0 0.9592 27.0 0.9231 45.0 0.8696 25.0 0.1154 46.0 0.7234 28.0 0.963 44.0 0.0 45.0 0.4255 31.0 0.0667 46.0 0.5 26.0 0.52 26.0 0.2222 94.0 0.3474 89.0 0.8462 44.0 0.0889 45.0 0.7021 27.0 0.1481 25.0 0.32 44.0 0.0217 29.0 0.0357 48.0 0.3958 47.0 0.7872 25.0 0.8077 27.0 0.1786 44.0 0.125 94.0 0.2947 28.0 0.3571 30.0 0.5 29.0 0.3793 28.0 0.25 44.0 0.2174 28.0 0.8214 26.0 0.64 28.0 0.8929 30.0 0.5517 27.0 0.1111 93.0 0.6809 26.0 0.72 42.0 0.9091 27.0 0.0 95.0 0.2245 44.0 0.4468 23.0 0.5417 28.0 0.963 28.0 1.0 49.0 0.1875 86.0 0.1573 44.0 0.617 31.0 0.7667 30.0 0.9655 28.0 0.8571 25.0 0.0 44.0 0.6522 29.0 0.2857 45.0 0.7609 27.0 0.0769 28.0 0.25 29.0 0.0714 47.0 0.9796 27.0 0.2963 28.0 0.5 50.0 0.5714 47.0 0.8958 30.0 0.2414 27.0 0.7143 25.0 0.3462 94.0 0.1489 25.0 0.2692 92.0 0.1075 48.0 0.2292 48.0 0.5 47.0 0.3958 20.0 0.25 26.0 0.4231 29.0 0.7586 47.0 0.5106 29.0 0.1071 88.0 0.9355 92.0 0.9891 29.0 0.1429 27.0 0.0 46.0 0.7021 27.0 0.6923 95.0 0.883 25.0 1.0 29.0 0.4643 89.0 0.0 46.0 0.0816 93.0 0.6526 28.0 0.3214 30.0 0.871 44.0 0.0833 27.0 0.2308 36.0 0.0 94.0 0.4105 47.0 1.0 44.0 0.3409 48.0 0.6327 94.0 0.3878 44.0 0.6667 28.0 1.0 27.0 0.5926 26.0 0.3462 93.0 0.6667 48.0 0.5625 27.0 0.037 25.0 0.68 27.0 0.8148 29.0 0.9643 49.0 0.4583 24.0 0.0385 46.0 0.6327 29.0 0.069 47.0 0.3542 96.0 0.3474 46.0 0.2083 97.0 0.0825 29.0 0.1786 96.0 0.7604 92.0 0.9255 27.0 0.0 29.0 0.1071 27.0 0.3846 32.0 0.2727 95.0 0.4421 25.0 0.037 28.0 0.1429 28.0 0.1071 41.0 0.0 47.0 0.6087 46.0 0.125 44.0 0.7333 48.0 0.3542 28.0 0.4074 85.0 0.9222 26.0 0.6538 45.0 0.1957 27.0 0.0357 49.0 0.5208 47.0 0.2292 29.0 0.2857 26.0 0.8889 50.0 0.9592 26.0 0.0769 27.0 0.9259 27.0 0.037 31.0 0.6774 48.0 0.5833 45.0 0.4222 47.0 0.9787 28.0 0.1786 30.0 0.8667 27.0 0.5 46.0 0.8085 93.0 0.1579 29.0 0.7333 27.0 0.0357 28.0 0.6296 30.0 0.8966 92.0 0.9375 28.0 0.1852 25.0 0.84 30.0 0.1034 97.0 1.0 44.0 0.0 27.0 0.1481 28.0 0.4138 25.0 0.04 27.0 0.8519 30.0 0.3448 91.0 0.4444 96.0 0.6082 93.0 0.9263 43.0 0.6591 26.0 0.5385 28.0 0.4444 26.0 0.9259 26.0 0.0741 49.0 0.4375 98.0 0.1313 44.0 0.0 28.0 0.037 27.0 0.6667 48.0 0.2128 27.0 0.3462 40.0 0.1111 93.0 0.1146 26.0 0.5385 97.0 0.5625 24.0 0.5 26.0 0.08 28.0 0.2963 28.0 0.6071 27.0 0.1111 45.0 0.3333 47.0 0.3125 25.0 0.2083 28.0 0.963 25.0 0.5385 25.0 0.24 26.0 0.1071 47.0 0.2857 29.0 0.7143 27.0 0.0769 44.0 0.413 89.0 0.2234 28.0 1.0 45.0 0.6889 42.0 0.122 48.0 0.1702 26.0 0.92 29.0 0.1379 23.0 0.0435 28.0 0.1111 26.0 0.0 47.0 0.2292 27.0 0.1538 27.0 0.7692 48.0 0.1702 25.0 0.1667 79.0 0.8925 26.0 0.6923 27.0 0.5185 27.0 0.1034 46.0 0.3404 26.0 0.1852 42.0 0.9091 48.0 0.25 30.0 0.4828 26.0 0.4444 45.0 0.2 27.0 0.3704 46.0 0.4255 27.0 0.7692 28.0 0.1429 31.0 0.8667 43.0 0.8889 28.0 0.8276 46.0 0.6739 27.0 0.9231 45.0 0.383 30.0 0.3448 16.0 0.8 46.0 0.9574 27.0 0.6429 27.0 0.5185 28.0 0.4815 31.0 0.7667 28.0 0.3704 31.0 0.8667 26.0 0.3462 98.0 0.8687 45.0 0.1224 46.0 0.5745 28.0 0.4815 26.0 0.6154 28.0 0.0714 25.0 0.8148 26.0 0.44 32.0 0.0606 47.0 0.1277 29.0 0.9 26.0 0.8148 97.0 0.9588 49.0 0.7708 29.0 0.5714 27.0 0.5769 27.0 0.3333 86.0 0.4205 27.0 0.0385 28.0 0.9643 92.0 0.8043 28.0 0.5714 47.0 0.0652 29.0 0.7857 95.0 0.9896 46.0 0.0833 29.0 1.0 28.0 0.0741 93.0 0.5263 24.0 0.0417 25.0 0.9259 27.0 0.9615 28.0 0.2963 27.0 0.3462 98.0 0.2653 26.0 0.48 26.0 0.037 28.0 0.5926 45.0 0.2917 30.0 0.7931 49.0 0.2857 45.0 0.3043 31.0 0.1562 32.0 0.4516 29.0 0.3103 28.0 0.2414 29.0 0.1379 29.0 0.7586 47.0 0.0851 26.0 0.1786 97.0 0.4227 27.0 0.0385 47.0 0.1875 49.0 0.6458 44.0 0.2128 30.0 0.7667 46.0 0.7447 27.0 0.8889 45.0 0.8298 96.0 0.5208 95.0 0.4947 48.0 0.6042 47.0 0.7551 46.0 0.8298 26.0 0.1852 46.0 0.766 43.0 0.5333 45.0 1.0 25.0 0.4167 49.0 0.2

In [0]:
%sql
select teamKills, winPlacePerc from project_dataset

teamKills winPlacePerc 0.0 0.4444 0.0 0.64 0.0 0.7755 0.0 0.1667 0.0 0.1875 0.0 0.037 0.0 0.0 0.0 0.7368 0.0 0.3704 0.0 0.2143 0.0 0.3929 0.0 0.4043 0.0 0.9286 0.0 0.875 0.0 0.9 0.0 0.2766 0.0 0.7308 0.0 0.8211 0.0 0.1923 0.0 0.931 0.0 0.6383 0.0 0.2143 0.0 0.75 0.0 0.9592 0.0 0.9231 0.0 0.8696 0.0 0.1154 0.0 0.7234 0.0 0.963 0.0 0.0 0.0 0.4255 0.0 0.0667 1.0 0.5 0.0 0.52 0.0 0.2222 0.0 0.3474 0.0 0.8462 0.0 0.0889 0.0 0.7021 0.0 0.1481 0.0 0.32 0.0 0.0217 0.0 0.0357 0.0 0.3958 0.0 0.7872 0.0 0.8077 0.0 0.1786 0.0 0.125 0.0 0.2947 0.0 0.3571 0.0 0.5 0.0 0.3793 0.0 0.25 0.0 0.2174 0.0 0.8214 0.0 0.64 0.0 0.8929 0.0 0.5517 0.0 0.1111 0.0 0.6809 0.0 0.72 0.0 0.9091 0.0 0.0 0.0 0.2245 0.0 0.4468 0.0 0.5417 0.0 0.963 0.0 1.0 0.0 0.1875 0.0 0.1573 0.0 0.617 0.0 0.7667 1.0 0.9655 0.0 0.8571 0.0 0.0 0.0 0.6522 0.0 0.2857 0.0 0.7609 0.0 0.0769 0.0 0.25 0.0 0.0714 0.0 0.9796 0.0 0.2963 0.0 0.5 0.0 0.5714 0.0 0.8958 0.0 0.2414 0.0 0.7143 0.0 0.3462 0.0 0.1489 0.0 0.2692 0.0 0.1075 0.0 0.2292 0.0 0.5 0.0 0.3958 0.0 0.25 0.0 0.4231 0.0 0.7586 0.0 0.5106 0.0 0.1071 0.0 0.9355 0.0 0.9891 0.0 0.1429 0.0 0.0 0.0 0.7021 0.0 0.6923 0.0 0.883 0.0 1.0 0.0 0.4643 0.0 0.0 0.0 0.0816 0.0 0.6526 0.0 0.3214 0.0 0.871 0.0 0.0833 0.0 0.2308 0.0 0.0 0.0 0.4105 0.0 1.0 0.0 0.3409 0.0 0.6327 0.0 0.3878 0.0 0.6667 0.0 1.0 0.0 0.5926 0.0 0.3462 0.0 0.6667 0.0 0.5625 0.0 0.037 0.0 0.68 0.0 0.8148 0.0 0.9643 0.0 0.4583 0.0 0.0385 0.0 0.6327 0.0 0.069 0.0 0.3542 0.0 0.3474 0.0 0.2083 0.0 0.0825 0.0 0.1786 0.0 0.7604 0.0 0.9255 0.0 0.0 0.0 0.1071 0.0 0.3846 0.0 0.2727 0.0 0.4421 0.0 0.037 0.0 0.1429 1.0 0.1071 0.0 0.0 0.0 0.6087 0.0 0.125 0.0 0.7333 0.0 0.3542 0.0 0.4074 0.0 0.9222 0.0 0.6538 0.0 0.1957 0.0 0.0357 0.0 0.5208 0.0 0.2292 0.0 0.2857 0.0 0.8889 0.0 0.9592 0.0 0.0769 0.0 0.9259 0.0 0.037 0.0 0.6774 0.0 0.5833 0.0 0.4222 0.0 0.9787 0.0 0.1786 0.0 0.8667 0.0 0.5 0.0 0.8085 0.0 0.1579 0.0 0.7333 0.0 0.0357 0.0 0.6296 0.0 0.8966 0.0 0.9375 0.0 0.1852 0.0 0.84 0.0 0.1034 0.0 1.0 0.0 0.0 0.0 0.1481 0.0 0.4138 0.0 0.04 0.0 0.8519 0.0 0.3448 0.0 0.4444 0.0 0.6082 0.0 0.9263 0.0 0.6591 0.0 0.5385 0.0 0.4444 1.0 0.9259 0.0 0.0741 0.0 0.4375 0.0 0.1313 0.0 0.0 0.0 0.037 0.0 0.6667 0.0 0.2128 0.0 0.3462 0.0 0.1111 0.0 0.1146 0.0 0.5385 0.0 0.5625 0.0 0.5 0.0 0.08 0.0 0.2963 0.0 0.6071 0.0 0.1111 0.0 0.3333 0.0 0.3125 0.0 0.2083 0.0 0.963 0.0 0.5385 0.0 0.24 0.0 0.1071 0.0 0.2857 0.0 0.7143 0.0 0.0769 0.0 0.413 0.0 0.2234 0.0 1.0 0.0 0.6889 0.0 0.122 0.0 0.1702 0.0 0.92 0.0 0.1379 0.0 0.0435 0.0 0.1111 0.0 0.0 0.0 0.2292 0.0 0.1538 0.0 0.7692 0.0 0.1702 0.0 0.1667 0.0 0.8925 0.0 0.6923 0.0 0.5185 0.0 0.1034 0.0 0.3404 0.0 0.1852 0.0 0.9091 0.0 0.25 0.0 0.4828 0.0 0.4444 0.0 0.2 0.0 0.3704 0.0 0.4255 1.0 0.7692 0.0 0.1429 0.0 0.8667 0.0 0.8889 0.0 0.8276 0.0 0.6739 0.0 0.9231 0.0 0.383 0.0 0.3448 0.0 0.8 0.0 0.9574 0.0 0.6429 0.0 0.5185 0.0 0.4815 0.0 0.7667 0.0 0.3704 0.0 0.8667 0.0 0.3462 0.0 0.8687 0.0 0.1224 0.0 0.5745 0.0 0.4815 0.0 0.6154 0.0 0.0714 0.0 0.8148 0.0 0.44 0.0 0.0606 0.0 0.1277 0.0 0.9 0.0 0.8148 0.0 0.9588 0.0 0.7708 0.0 0.5714 0.0 0.5769 0.0 0.3333 0.0 0.4205 0.0 0.0385 0.0 0.9643 0.0 0.8043 0.0 0.5714 0.0 0.0652 0.0 0.7857 0.0 0.9896 0.0 0.0833 0.0 1.0 0.0 0.0741 0.0 0.5263 0.0 0.0417 0.0 0.9259 0.0 0.9615 0.0 0.2963 0.0 0.3462 0.0 0.2653 0.0 0.48 0.0 0.037 0.0 0.5926 0.0 0.2917 0.0 0.7931 0.0 0.2857 0.0 0.3043 0.0 0.1562 0.0 0.4516 0.0 0.3103 0.0 0.2414 0.0 0.1379 0.0 0.7586 0.0 0.0851 0.0 0.1786 0.0 0.4227 0.0 0.0385 0.0 0.1875 0.0 0.6458 0.0 0.2128 1.0 0.7667 0.0 0.7447 0.0 0.8889 0.0 0.8298 0.0 0.5208 0.0 0.4947 0.0 0.6042 0.0 0.7551 0.0 0.8298 0.0 0.1852 0.0 0.766 0.0 0.5333 0.0 1.0 0.0 0.4167 0.0 0.2083 0.0 0.8958 0.0 0.4615 0.0 0.931 0.0 0.2143 0.0 0.16 0.0 0.7312 0.0 0.5172 0.0 0.12 0.0 0.8462 0.0 0.0 0.0 0.0769 0.0 0.9655 0.0 0.8958 0.0 0.0741 0.0 0.0 0.0 0.1333 0.0 0.4615 0.0 0.5979 0.0 0.4194 0.0 0.9286 0.0 0.5333 0.0 0.3929 0.0 0.5312 0.0 0.3704 0.0 0.9167 0.0 0.9231 0.0 0.037 0.0 0.8511 0.0 0.2963 0.0 0.3214 0.0 0.4545 0.0 0.32 0.

**As a conclusion, we find that matchDuration, maxPlace, numGroups, teamKills have little effect on our target, but walkDistance and killPlace has a certain effect. So we need to drop these three columns to decrease the number of calculations**

In [0]:
trainDF = trainDF.drop('matchDuration', 'maxPlace', 'numGroups', 'teamKills')

## Drop rows with null value
trainDF = trainDF.na.drop()

sqlContext.sql("DROP TABLE IF EXISTS project_dataset")
dbutils.fs.rm("dbfs:/user/hive/warehouse/project_dataset", True)
sqlContext.registerDataFrameAsTable(trainDF, "project_dataset")

display(trainDF)

assists boosts damageDealt DBNOs headshotKills heals killPlace kills killStreaks revives rideDistance roadKills swimDistance vehicleDestroys walkDistance weaponsAcquired winPlacePerc 0.0 0.0 0.0 0.0 0.0 0.0 60.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 244.8 1.0 0.4444 0.0 0.0 91.47 0.0 0.0 0.0 57.0 0.0 0.0 0.0 0.0045 0.0 11.04 0.0 1434.0 5.0 0.64 1.0 0.0 68.0 0.0 0.0 0.0 47.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 161.8 2.0 0.7755 0.0 0.0 32.9 0.0 0.0 0.0 75.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 202.7 3.0 0.1667 0.0 0.0 100.0 0.0 0.0 0.0 45.0 1.0 1.0 0.0 0.0 0.0 0.0 0.0 49.75 2.0 0.1875 0.0 0.0 100.0 1.0 1.0 0.0 44.0 1.0 1.0 0.0 0.0 0.0 0.0 0.0 34.7 1.0 0.037 0.0 0.0 0.0 0.0 0.0 0.0 96.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 13.5 1.0 0.0 0.0 0.0 8.538 0.0 0.0 0.0 48.0 0.0 0.0 0.0 2004.0 0.0 0.0 0.0 1089.0 6.0 0.7368 0.0 0.0 51.6 0.0 0.0 0.0 64.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 799.9 4.0 0.3704 0.0 0.0 37.27 0.0 0.0 0.0 74.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 65.67 1.0 0.2143 0.0 0.0 28.38 0.0 0.0 0.0 75.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 868.3 9.0 0.3929 0.0 0.0 137.9 1.0 0.0 0.0 64.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 451.7 1.0 0.4043 0.0 0.0 0.0 0.0 0.0 0.0 37.0 0.0 0.0 0.0 6639.0 0.0 0.0 0.0 2784.0 6.0 0.9286 0.0 1.0 324.2 0.0 1.0 5.0 5.0 4.0 1.0 0.0 1228.0 0.0 76.84 0.0 2050.0 6.0 0.875 0.0 1.0 122.8 1.0 0.0 2.0 25.0 1.0 1.0 1.0 1237.0 0.0 60.29 0.0 1666.0 5.0 0.9 0.0 0.0 80.71 1.0 0.0 0.0 72.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 105.1 5.0 0.2766 0.0 2.0 81.71 1.0 0.0 14.0 25.0 1.0 1.0 0.0 519.9 0.0 0.0 0.0 3674.0 7.0 0.7308 0.0 3.0 254.3 0.0 0.0 12.0 13.0 2.0 1.0 0.0 2367.0 0.0 15.29 0.0 1787.0 3.0 0.8211 0.0 0.0 0.0 0.0 0.0 0.0 79.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 137.4 2.0 0.1923 0.0 1.0 65.28 0.0 0.0 1.0 48.0 0.0 0.0 0.0 0.0 0.0 20.85 0.0 3310.0 3.0 0.931 0.0 4.0 269.1 0.0 1.0 8.0 18.0 2.0 1.0 1.0 2734.0 0.0 0.0 0.0 1794.0 5.0 0.6383 0.0 0.0 158.7 1.0 0.0 0.0 75.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 580.1 2.0 0.2143 0.0 1.0 192.3 1.0 2.0 3.0 15.0 2.0 1.0 0.0 2332.0 0.0 0.0 0.0 1264.0 4.0 0.75 0.0 6.0 1011.0 6.0 2.0 2.0 2.0 9.0 2.0 0.0 4860.0 0.0 0.0 0.0 2727.0 7.0 0.9592 0.0 3.0 327.6 4.0 1.0 1.0 3.0 4.0 2.0 0.0 0.0 0.0 0.0 0.0 3503.0 4.0 0.9231 1.0 4.0 558.6 3.0 0.0 4.0 11.0 3.0 2.0 0.0 1183.0 0.0 0.0 0.0 2711.0 7.0 0.8696 0.0 0.0 44.28 0.0 0.0 0.0 78.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 15.3 0.0 0.1154 0.0 4.0 381.2 2.0 1.0 2.0 7.0 4.0 2.0 0.0 1798.0 0.0 0.0 0.0 1933.0 4.0 0.7234 0.0 6.0 345.6 2.0 1.0 1.0 6.0 4.0 1.0 0.0 0.0 0.0 0.0 0.0 3855.0 4.0 0.963 0.0 0.0 0.0 0.0 0.0 0.0 87.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 36.73 0.0 0.0 0.0 62.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 455.5 5.0 0.4255 0.0 0.0 0.0 0.0 0.0 0.0 80.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 93.11 2.0 0.0667 0.0 3.0 0.0 0.0 0.0 1.0 61.0 0.0 0.0 0.0 1154.0 0.0 0.0 0.0 1240.0 7.0 0.5 0.0 2.0 65.47 0.0 0.0 2.0 61.0 0.0 0.0 3.0 0.0 0.0 0.0 0.0 751.8 8.0 0.52 0.0 1.0 439.9 4.0 0.0 1.0 34.0 1.0 1.0 0.0 0.0 0.0 0.0 0.0 353.5 2.0 0.2222 0.0 0.0 136.9 0.0 0.0 0.0 37.0 1.0 1.0 0.0 0.0 0.0 0.0 0.0 270.7 1.0 0.3474 0.0 1.0 100.0 0.0 0.0 0.0 24.0 1.0 1.0 0.0 0.0 0.0 61.0 0.0 1528.0 3.0 0.8462 0.0 0.0 0.0 0.0 0.0 0.0 82.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 60.93 1.0 0.0889 1.0 0.0 318.0 2.0 0.0 0.0 5.0 4.0 2.0 0.0 0.0 0.0 0.0 0.0 2303.0 2.0 0.7021 1.0 0.0 177.8 1.0 0.0 0.0 73.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 133.4 1.0 0.1481 0.0 1.0 61.9 1.0 0.0 1.0 31.0 1.0 1.0 0.0 529.1 0.0 0.0 0.0 327.3 2.0 0.32 0.0 0.0 0.0 0.0 0.0 0.0 86.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 128.8 1.0 0.0217 0.0 0.0 245.6 2.0 1.0 0.0 46.0 1.0 1.0 0.0 0.0 0.0 0.0 0.0 52.52 1.0 0.0357 0.0 1.0 529.3 3.0 0.0 0.0 12.0 3.0 2.0 0.0 0.0 0.0 0.0 0.0 534.1 4.0 0.3958 0.0 4.0 401.2 3.0 0.0 6.0 27.0 1.0 1.0 1.0 0.0 0.0 0.0 0.0 2576.0 4.0 0.7872 3.0 1.0 395.2 2.0 1.0 2.0 13.0 2.0 1.0 0.0 0.0 0.0 0.0 0.0 2003.0 3.0 0.8077 0.0 0.0 0.0 0.0 0.0 0.0 77.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 68.56 2.0 0.1786 1.0 0.0 0.0 0.0 0.0 0.0 79.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 79.33 3.0 0.125 0.0 0.0 194.5 0.0 1.0 1.0 19.0 2.0 1.0 0.0 0.0 0.0 0.0 0.0 248.6 5.0 0.2947 0.0 2.0 279.0 4.0 0.0 1.0 10.0 3.0 1.0 1.0 0.0 0.0 0.0 0.0 991.7 2.0 0.3571 0.0 0

**Now we can check whether there is a null value in the dataset. If there is one null value, we have to drop this row.**
 - Since we have handled the null values in some rows before, it should have no outputs.

In [0]:
%sql
select winPlacePerc from project_dataset
where winPlacePerc is null

winPlacePerc

##Part 3: Data Preparation

Then we need to prepare the data for machine learning. 

- Convert the `project_dataset` SQL table into a DataFrame.
- Set the vectorizer's input columns to a list of the sixteen columns of the input DataFrame.
- Set the vectorizer's output column name to `"features"`.

In [0]:
from pyspark.ml.feature import VectorAssembler

datasetDF = sqlContext.table("project_dataset")

vectorizer = VectorAssembler()
vectorizer.setInputCols(["assists", "boosts", "damageDealt", "DBNOs", "headshotKills", "heals", "killPlace", "kills", "killStreaks", "revives", "rideDistance", "roadKills", "swimDistance", "vehicleDestroys", "walkDistance", "weaponsAcquired"])
vectorizer.setOutputCol("features")

Out[123]: VectorAssembler_3be65ebfcc20

**In order to see how well the model is, we need to split the dataset into training set and test set. Then we cache them.**

In [0]:
(split15DF, split85DF) = datasetDF.randomSplit([0.15, 0.85], seed = "20583761")
testSetDF = split15DF.cache()
trainingSetDF = split85DF.cache()

**Before we apply any of the algorithm, we first generate linear analyses of the dataset to see some preliminary results**

In [0]:
from pyspark.ml.regression import LinearRegression
from pyspark.ml.regression import LinearRegressionModel
from pyspark.ml import Pipeline

lr = LinearRegression()
lr.setPredictionCol("Prediction_WP")\
  .setLabelCol("winPlacePerc")\
  .setMaxIter(100)\
  .setRegParam(0.15)

lrPipeline = Pipeline()
lrPipeline.setStages([vectorizer, lr])
lrModel = lrPipeline.fit(trainingSetDF)

**Then we can draw the equation of the linear regrassion to see a preliminary results.**

In [0]:
# The intercept is as follows:
intercept = lrModel.stages[1].intercept

# The coefficents (i.e., weights) are as follows:
weights = lrModel.stages[1].coefficients

# Create a list of the column names (without PE)
featuresNoLabel = [col for col in datasetDF.columns if col != "winPlacePerc"]

# Merge the weights and labels
coefficents = zip(weights, featuresNoLabel)

# Now let's sort the coefficients from greatest absolute weight most to the least absolute weight
coefficents.sort(key=lambda tup: abs(tup[0]), reverse=True)

equation = "y = {intercept}".format(intercept=intercept)
variables = []
for x in coefficents:
    weight = abs(x[0])
    name = x[1]
    symbol = "+" if (x[0] > 0) else "-"
    equation += (" {} ({} * {})".format(symbol, weight, name))

# Finally here is our equation
print("Linear Regression Equation: " + equation)

Linear Regression Equation: y = 0.413646647358 + (0.0230870694198 * boosts) + (0.0179682771692 * assists) + (0.0171856868531 * weaponsAcquired) + (0.014296475364 * revives) - (0.0124608091795 * killStreaks) - (0.0107073646573 * DBNOs) + (0.00597216525247 * roadKills) + (0.00481247690183 * heals) - (0.00284673719129 * killPlace) + (0.00186405039592 * vehicleDestroys) + (0.00127669784278 * headshotKills) + (0.00030038312258 * swimDistance) - (0.00013574919577 * kills) + (7.99806962651e-05 * walkDistance) + (4.24414949638e-05 * damageDealt) + (1.22938354063e-05 * rideDistance)

**Next we apply the Linear Regression model to the 15% of the data that we split from the input dataset. The output of the model will be a predicted win place named "Prediction_WP". Then we calculate its rmse and \\(r^2\\)**

In [0]:
resultsDF = lrModel.transform(testSetDF).select("assists", "boosts", "damageDealt", "DBNOs", "headshotKills", "heals", "killPlace", "kills", "killStreaks", "revives", "rideDistance", "roadKills", "swimDistance", "vehicleDestroys", "walkDistance", "weaponsAcquired", "winPlacePerc", "Prediction_WP")
display(resultsDF)

assists boosts damageDealt DBNOs headshotKills heals killPlace kills killStreaks revives rideDistance roadKills swimDistance vehicleDestroys walkDistance weaponsAcquired winPlacePerc Prediction_WP 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 1.0 0.44517128387325133 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.5482854049916153 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.40795317297584255 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 1.0 0.4251388598289032 0.0 0.0 0.0 0.0 0.0 0.0 3.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 9.0 1.0 0.559777617462101 0.0 0.0 0.0 0.0 0.0 0.0 4.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 4.0 0.0 0.4710024460055102 0.0 0.0 0.0 0.0 0.0 0.0 4.0 0.0 0.0 0.0 7302.0 0.0 4.851 0.0 592.9 5.0 1.0 0.6268354323388641 0.0 0.0 0.0 0.0 0.0 0.0 5.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.43378433510810144 0.0 0.0 0.0 0.0 0.0 0.0 5.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 5.0 0.0 0.4853413956672834 0.0 0.0 0.0 0.0 0.0 0.0 5.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 6.0 0.0 0.5025270825203441 0.0 0.0 0.0 0.0 0.0 0.0 5.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 9.0 1.0 0.554084143079526 0.0 0.0 0.0 0.0 0.0 0.0 6.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 3.0 0.0 0.4481232847698746 0.0 0.0 0.0 0.0 0.0 0.0 6.0 0.0 0.0 0.0 362.0 0.0 0.0 0.0 533.6 6.0 0.0 0.5468084132732114 0.0 0.0 0.0 0.0 0.0 0.0 7.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 3.0 0.0 0.44527654757858715 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.4080584366811783 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.42524412353423896 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 3.0 0.0 0.44242981038729967 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 4.0 0.0 0.4596154972403603 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 6.0 0.0 0.49398687094648164 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.4223973863429515 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 5.0 0.0 0.4739544469021335 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.5255115074613155 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.5426971943143761 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.5598828811674368 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.5598828811674368 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.40236496229860336 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.40236496229860336 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.419550649151664 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 4.0 0.0 0.4539220228577853 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 4.0 0.0 0.4539220228577853 0.0 0.0 0.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 9.0 0.0 0.5398504571230887 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.3823325382542552 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.3995182251073159 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.3995182251073159 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.41670391196037654 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.41670391196037654 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 7.0 0.0 0.5026323462256799 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.5198180330787405 0.0 0.0 0.0 0.0 0.0 0.0 11.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 11.25 3.0 0.4286 0.43478938164641967 0.0 0.0 0.0 0.0 0.0 0.0 12.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 13.0 0.0 0.6028997301527563 0.0 0.0 0.0 0.0 0.0 0.0 12.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2221.0 3.0 0.0 0.60867998802695 0.0 0.0 0.0 0.0 0.0 0.0 13.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 7.0 0.0 0.4969388718431049 0.0 0.0 0.0 0.0 0.0 0.0 13.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 8.0 0.0 0.5141245586961656 0.0 0.0 0.0 0.0

In [0]:
from pyspark.ml.evaluation import RegressionEvaluator

regEval = RegressionEvaluator(predictionCol="Prediction_WP", labelCol="winPlacePerc", metricName="rmse")
rmse = regEval.evaluate(resultsDF)
r2 = regEval.evaluate(resultsDF, {regEval.metricName: "r2"})

print("Root Mean Squared Error: %.2f" % rmse)
print("r2: {0:.2f}".format(r2))

Root Mean Squared Error: 0.16
r2: 0.74

**Then we can construct the pie chart of RMSE to see whether a linear regression is perfect.**

In [0]:
sqlContext.sql("DROP TABLE IF EXISTS Project_RMSE_Evaluation")
dbutils.fs.rm("dbfs:/user/hive/warehouse/Project_RMSE_Evaluation", True)

# Next we calculate the residual error and divide it by the RMSE
resultsDF.selectExpr("winPlacePerc", "Prediction_WP", "winPlacePerc - Prediction_WP Residual_Error", "(winPlacePerc - Prediction_WP) / {} Within_RSME".format(rmse)).registerTempTable("Project_RMSE_Evaluation")

In [0]:
%sql
SELECT case 
       when Within_RSME <= 1.0 AND Within_RSME >= -1.0 then 1
       when Within_RSME <= 2.0 AND Within_RSME >= -2.0 then 2 
       else 3
       end RSME_Multiple, COUNT(*) AS count
FROM Project_RMSE_Evaluation
GROUP BY RSME_Multiple

RSME_Multiple count 1 462875 3 26822 2 177749

From the pie chart, we can see that 69% of our test data predictions are within 1 RMSE of the actual values, and 96% (69% + 27%) of our test data predictions are within 2 RMSE. So the model is pretty decent. Then we can use tuning to improve it.

##Part 4: Evaluation with Tuning

To improve the results, we can directly use decision tree and random forest to get the improvement. But firstly we can use the decision tree to see its accuracy. Then we use the random forest to see whether there can be more accuracy.

In [0]:
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator
from pyspark.ml.regression import DecisionTreeRegressor

# Create a DecisionTreeRegressor
dt = DecisionTreeRegressor()
dt.setPredictionCol("Prediction_WP").setFeaturesCol("features").setLabelCol("winPlacePerc").setMaxBins(100)

# Create a Pipeline
dtPipeline = Pipeline()

# Set the stages of the Pipeline
dtPipeline.setStages([vectorizer, dt])

# Reuse the CrossValidator
crossval = CrossValidator(estimator=dtPipeline, evaluator=regEval, numFolds=3)

# Create a paramter grid using the ParamGridBuilder
paramGrid = (ParamGridBuilder()
             .addGrid(dt.maxDepth, [14, 15])
             .build())

# Add the grid to the CrossValidator
crossval.setEstimatorParamMaps(paramGrid)

# Find and return the best model
dtModel = crossval.fit(trainingSetDF).bestModel

MLlib will automatically track trials in MLflow. After your tuning fit() call has completed, view the MLflow UI to see logged runs.

**We can evaluate the model by calculate DT model's RMSE and \\(r^2\\) values.**

In [0]:
resultsDF = dtModel.transform(testSetDF).select("assists", "boosts", "damageDealt", "DBNOs", "headshotKills", "heals", "killPlace", "kills", "killStreaks", "revives", "rideDistance", "roadKills", "swimDistance", "vehicleDestroys", "walkDistance", "weaponsAcquired", "winPlacePerc", "Prediction_WP")
rmseDT = regEval.evaluate(resultsDF)
r2DT = regEval.evaluate(resultsDF, {regEval.metricName: "r2"})

print("DT Root Mean Squared Error: {0:.2f}".format(rmseDT))
print("DT r2: {0:.2f}".format(r2DT))

DT Root Mean Squared Error: 0.10
DT r2: 0.89

The line below will pull the Decision Tree model from the Pipeline as display it as an if-then-else string.

In [0]:
print dtModel.stages[-1]._java_obj.toDebugString()

DecisionTreeRegressionModel (uid=DecisionTreeRegressor_57e1bf8f9dd4) of depth 14 with 25903 nodes
 If (feature 14 <= 847.25)
 If (feature 6 <= 74.5)
 If (feature 14 <= 306.45)
 If (feature 7 <= 0.5)
 If (feature 6 <= 62.5)
 If (feature 6 <= 29.5)
 If (feature 6 <= 3.5)
 If (feature 6 <= 1.5)
 If (feature 2 <= 27.6)
 If (feature 15 <= 9.5)
 If (feature 1 <= 0.5)
 If (feature 10 <= 7.0E-4)
 If (feature 14 <= 5.894)
 If (feature 15 <= 3.5)
 Predict: 0.813953488372093
 Else (feature 15 > 3.5)
 Predict: 0.7209302325581395
 Else (feature 14 > 5.894)
 Predict: 0.0
 Else (feature 10 > 7.0E-4)
 Predict: 1.0
 Else (feature 1 > 0.5)
 Predict: 1.0
 Else (feature 15 > 9.5)
 If (feature 1 <= 0.5)
 If (feature 5 <= 0.5)
 If (feature 2 <= 0.0796)
 If (feature 15 <= 12.5)
 Predict: 0.375
 Else (feature 15 > 12.5)
 Predict: 0.6666666666666666
 Else (feature 2 > 0.0796)
 Predict: 1.0
 Else (feature 5 > 0.5)
 Predict: 0.0
 Else (feature 1 > 0.5)
 If (feature 15 <= 13.5)
 Predict: 1.0
 Else (feature 15 > 13.5)
 If (feature 1 <= 1.5)
 Predict: 0.5
 Else (feature 1 > 1.5)
 Predict: 1.0
 Else (feature 2 > 27.6)
 If (feature 5 <= 7.5)
 If (feature 5 <= 0.5)
 If (feature 1 <= 1.5)
 If (feature 2 <= 131.05)
 If (feature 2 <= 81.005)
 Predict: 0.8387096774193549
 Else (feature 2 > 81.005)
 Predict: 0.9782608695652174
 Else (feature 2 > 131.05)
 If (feature 2 <= 137.75)
 Predict: 0.0
 Else (feature 2 > 137.75)
 Predict: 0.8414634146341463
 Else (feature 1 > 1.5)
 Predict: 0.0
 Else (feature 5 > 0.5)
 Predict: 1.0
 Else (feature 5 > 7.5)
 Predict: 0.0
 Else (feature 6 > 1.5)
 If (feature 6 <= 2.5)
 If (feature 15 <= 15.5)
 If (feature 2 <= 222.05)
 If (feature 2 <= 88.975)
 If (feature 2 <= 66.935)
 If (feature 2 <= 25.335)
 Predict: 0.6260162601626016
 Else (feature 2 > 25.335)
 Predict: 0.40625
 Else (feature 2 > 66.935)
 If (feature 1 <= 0.5)
 Predict: 0.8823529411764706
 Else (feature 1 > 0.5)
 Predict: 0.0
 Else (feature 2 > 88.975)
 If (feature 2 <= 93.16499999999999)
 Predict: 0.0
 Else (feature 2 > 93.16499999999999)
 If (feature 14 <= 1.0E-4)
 Predict: 0.5375
 Else (feature 14 > 1.0E-4)
 Predict: 0.16666666666666666
 Else (feature 2 > 222.05)
 If (feature 5 <= 0.5)
 If (feature 1 <= 3.5)
 If (feature 15 <= 11.5)
 Predict: 0.717948717948718
 Else (feature 15 > 11.5)
 Predict: 1.0
 Else (feature 1 > 3.5)
 Predict: 0.0
 Else (feature 5 > 0.5)
 If (feature 1 <= 1.5)
 If (feature 15 <= 7.5)
 Predict: 0.0
 Else (feature 15 > 7.5)
 Predict: 0.5
 Else (feature 1 > 1.5)
 Predict: 1.0
 Else (feature 15 > 15.5)
 If (feature 2 <= 52.43)
 If (feature 2 <= 19.365000000000002)
 If (feature 15 <= 18.0)
 Predict: 1.0
 Else (feature 15 > 18.0)
 If (feature 15 <= 19.5)
 Predict: 0.0
 Else (feature 15 > 19.5)
 Predict: 0.75
 Else (feature 2 > 19.365000000000002)
 Predict: 0.0
 Else (feature 2 > 52.43)
 If (feature 2 <= 99.97)
 Predict: 1.0
 Else (feature 2 > 99.97)
 If (feature 2 <= 166.25)
 Predict: 0.0
 Else (feature 2 > 166.25)
 Predict: 1.0
 Else (feature 6 > 2.5)
 If (feature 15 <= 11.5)
 If (feature 14 <= 225.95)
 If (feature 2 <= 131.05)
 If (feature 2 <= 99.97)
 If (feature 2 <= 0.0796)
 Predict: 0.42
 Else (feature 2 > 0.0796)
 Predict: 0.3
 Else (feature 2 > 99.97)
 If (feature 2 <= 105.75)
 Predict: 1.0
 Else (feature 2 > 105.75)
 Predict: 0.5
 Else (feature 2 > 131.05)
 If (feature 15 <= 4.5)
 If (feature 1 <= 1.5)
 Predict: 0.3333333333333333
 Else (feature 1 > 1.5)
 Predict: 1.0
 Else (feature 15 > 4.5)
 If (feature 2 <= 299.95)
 Predict: 0.19444444444444445
 Else (feature 2 > 299.95)
 Predict: 0.0
 Else (feature 14 > 225.95)
 Predict: 1.0
 Else (feature 15 > 11.5)
 If (feature 2 <= 222.05)
 If (feature 2 <= 88.975)
 If (feature 15 <= 14.5)
 Predict: 1.0
 Else (feature 15 > 14.5)
 If (feature 15 <= 15.5)
 Predict: 0.3333333333333333
 Else (feature 15 > 15.5)
 Predict: 0.875
 Else (feature 2 > 88.975)
 If (feature 1 <= 0.5)
 If (feature 2 <= 105.75)
 Predict: 0.0
 Else (feature 2 > 105.75)
 Predict: 0.5
 Else (feature 1 > 0.5)
 Predict: 1.0
 Else (featur

**Notice that the depth of dt we estimated is 14 or 15 which is slightly small. So we have to adopt random tree to see whether it can be improved. And instead of guessing the parameters, we use Model Selection or Hyperparameter Tuning to select the best model**

The parameters for the method list below:

 - Set the name of the prediction column to "Prediction_WP"
 - Set the name of the label column to "winPlacePerc"
 - Set the name of the features column to "features"
 - Set the random number generator seed to 20583761
 - Set the maximum depth to 10
 - Set the number of trees to 24

In [0]:
from pyspark.ml.regression import RandomForestRegressor

# Create a RandomForestRegressor
rf = RandomForestRegressor()
rf.setPredictionCol("Prediction_WP").setLabelCol("winPlacePerc").setFeaturesCol("features").setSeed(20583761).setMaxDepth(10).setNumTrees(24)

# Create a Pipeline and set the stages of the Pipeline
rfPipeline = Pipeline()
rfPipeline.setStages([vectorizer, rf])

# Reuse the CrossValidator
crossval.setEstimator(rfPipeline)

# Tune over rf.maxBins parameter on the values 50 and 100, create a paramter grid using the ParamGridBuilder
paramGrid = ParamGridBuilder().addGrid(rf.maxBins, [50, 100]).build()

# Add the grid to the CrossValidator
crossval.setEstimatorParamMaps(paramGrid)

# Find and return the best model
rfModel = crossval.fit(trainingSetDF).bestModel

MLlib will automatically track trials in MLflow. After your tuning fit() call has completed, view the MLflow UI to see logged runs.

Now we can get the tuned RandomForestRegressor model's RMSE and \\(r^2\\) values and compare them to our tuned DecisionTreeRegressor models.

In [0]:
resultsDF = rfModel.transform(testSetDF).select("assists", "boosts", "damageDealt", "DBNOs", "headshotKills", "heals", "killPlace", "kills", "killStreaks", "revives", "rideDistance", "roadKills", "swimDistance", "vehicleDestroys", "walkDistance", "weaponsAcquired", "winPlacePerc", "Prediction_WP")
rmseRF = regEval.evaluate(resultsDF)
r2RF = regEval.evaluate(resultsDF, {regEval.metricName: "r2"})

print("LR Root Mean Squared Error: {0:.2f}".format(rmse))
print("DT Root Mean Squared Error: {0:.2f}".format(rmseDT))
print("RF Root Mean Squared Error: {0:.2f}".format(rmseRF))
print("LR r2: {0:.2f}".format(r2))
print("DT r2: {0:.2f}".format(r2DT))
print("RF r2: {0:.2f}".format(r2RF))

LR Root Mean Squared Error: 0.16
DT Root Mean Squared Error: 0.10
RF Root Mean Squared Error: 0.11
LR r2: 0.74
DT r2: 0.89
RF r2: 0.88

Unfortunately, the performance of random tree is not as good as decision tree. But we still get an optimal result with relatively low RMSE and high \\(r^2\\). Then we pull the Random Forest model from the Pipeline as an if-then-else string.

In [0]:
print rfModel.stages[-1]._java_obj.toDebugString()

RandomForestRegressionModel (uid=RandomForestRegressor_9c77a9d4daa0) with 24 trees
 Tree 0 (weight 1.0):
 If (feature 14 <= 846.4)
 If (feature 6 <= 74.5)
 If (feature 15 <= 2.5)
 If (feature 14 <= 269.25)
 If (feature 8 <= 0.5)
 If (feature 0 <= 0.5)
 If (feature 15 <= 1.5)
 If (feature 6 <= 60.5)
 If (feature 14 <= 0.06665)
 If (feature 6 <= 40.5)
 Predict: 0.05531861341371514
 Else (feature 6 > 40.5)
 Predict: 0.5017071591774562
 Else (feature 14 > 0.06665)
 If (feature 6 <= 52.5)
 Predict: 0.7075328283970156
 Else (feature 6 > 52.5)
 Predict: 0.5037653100775193
 Else (feature 6 > 60.5)
 If (feature 6 <= 68.5)
 If (feature 6 <= 64.5)
 Predict: 0.391635298804781
 Else (feature 6 > 64.5)
 Predict: 0.3278795935707712
 Else (feature 6 > 68.5)
 If (feature 6 <= 71.5)
 Predict: 0.27492044796292725
 Else (feature 6 > 71.5)
 Predict: 0.23063099945085105
 Else (feature 15 > 1.5)
 If (feature 6 <= 62.5)
 If (feature 14 <= 0.06665)
 If (feature 5 <= 0.5)
 Predict: 0.0917842064112588
 Else (feature 5 > 0.5)
 Predict: 0.2
 Else (feature 14 > 0.06665)
 If (feature 5 <= 0.5)
 Predict: 0.5379925179312944
 Else (feature 5 > 0.5)
 Predict: 0.49370305383022844
 Else (feature 6 > 62.5)
 If (feature 1 <= 0.5)
 If (feature 6 <= 69.5)
 Predict: 0.33380500666370505
 Else (feature 6 > 69.5)
 Predict: 0.2531989354406819
 Else (feature 1 > 0.5)
 If (feature 5 <= 0.5)
 Predict: 0.29794509915014167
 Else (feature 5 > 0.5)
 Predict: 0.3141862068965517
 Else (feature 0 > 0.5)
 If (feature 6 <= 57.5)
 If (feature 6 <= 50.5)
 If (feature 2 <= 182.75)
 If (feature 6 <= 35.5)
 Predict: 0.42914655172413785
 Else (feature 6 > 35.5)
 Predict: 0.8094916037735849
 Else (feature 2 > 182.75)
 If (feature 6 <= 40.5)
 Predict: 0.36974999999999997
 Else (feature 6 > 40.5)
 Predict: 0.7161103448275863
 Else (feature 6 > 50.5)
 If (feature 6 <= 54.5)
 If (feature 15 <= 0.5)
 Predict: 0.48478695652173914
 Else (feature 15 > 0.5)
 Predict: 0.6296347252747253
 Else (feature 6 > 54.5)
 If (feature 15 <= 1.5)
 Predict: 0.5106298507462687
 Else (feature 15 > 1.5)
 Predict: 0.52913536121673
 Else (feature 6 > 57.5)
 If (feature 14 <= 199.7)
 If (feature 3 <= 0.5)
 If (feature 12 <= 2.4295)
 Predict: 0.3098684074299636
 Else (feature 12 > 2.4295)
 Predict: 0.38710000000000594
 Else (feature 3 > 0.5)
 If (feature 3 <= 1.5)
 Predict: 0.2952900398406374
 Else (feature 3 > 1.5)
 Predict: 0.26082394366197165
 Else (feature 14 > 199.7)
 If (feature 15 <= 1.5)
 If (feature 5 <= 2.5)
 Predict: 0.31254041204437405
 Else (feature 5 > 2.5)
 Predict: 0.3643999999999994
 Else (feature 15 > 1.5)
 If (feature 6 <= 65.5)
 Predict: 0.4047428828828829
 Else (feature 6 > 65.5)
 Predict: 0.2987481679389314
 Else (feature 8 > 0.5)
 If (feature 1 <= 0.5)
 If (feature 14 <= 109.25)
 If (feature 6 <= 35.5)
 If (feature 7 <= 1.5)
 If (feature 6 <= 28.5)
 Predict: 0.6292303236797273
 Else (feature 6 > 28.5)
 Predict: 0.24320003229713363
 Else (feature 7 > 1.5)
 If (feature 10 <= 4520.5)
 Predict: 0.17084232568626326
 Else (feature 10 > 4520.5)
 Predict: 0.7323964285714217
 Else (feature 6 > 35.5)
 If (feature 6 <= 40.5)
 If (feature 3 <= 0.5)
 Predict: 0.17664336451330015
 Else (feature 3 > 0.5)
 Predict: 0.12443423478526779
 Else (feature 6 > 40.5)
 If (feature 15 <= 1.5)
 Predict: 0.08860522453258764
 Else (feature 15 > 1.5)
 Predict: 0.11883269096005585
 Else (feature 14 > 109.25)
 If (feature 14 <= 177.3)
 If (feature 6 <= 36.5)
 If (feature 8 <= 1.5)
 Predict: 0.2659517767653759
 Else (feature 8 > 1.5)
 Predict: 0.2140026942355888
 Else (feature 6 > 36.5)
 If (feature 14 <= 144.2)
 Predict: 0.16690389349020823
 Else (feature 14 > 144.2)
 Predict: 0.182164039080715
 Else (feature 14 > 177.3)
 If (feature 11 <= 1.5)
 If (feature 2 <= 201.35000000000002)
 Predict: 0.2527208922625227
 Else (feature 2 > 201.35000000000002)
 Predict: 0.27643696962243264
 Else (feature 11 > 1.5)
 If (feature 10 <= 2140.5)
 Predict: 0.244745
 Else (feature 10 > 2140.5)
 Predict: 0.8084333333333333
 Else (feature 1 > 0.5)
 

## Part 5: Conclusion

Till now, we have generated three kinds of models to evaluate the winPlacePerc, which represents of win place percentage. For the last two models, we can find that feature 6(killPlace) and feature 14(walkDistance) has a huge effect on our target. But considering the fact that the person who gets the champion is more likely to have a high killPlace and walk a long distance, we can look at the linear regression expression to get some ideas:

`Linear Regression Equation: y = 0.413646647358 + (0.0230870694198 * boosts) + (0.0179682771692 * assists) + ...`

From the equation, we notice that boosts, assists and weaponsAcquire have a high coefficient. We may get some conclusions saying that more boosts you throw, more assists you get, and more weapons you pick, you may have more chance to get the champion and eat the chicken!

In the end, we apply our best model of decision tree to generate the results of the testSet given by the websites.

In [0]:
# File location and type
file_location = "/FileStore/tables/train_V2-2.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
testDF = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(testDF)

Id groupId matchId assists boosts damageDealt DBNOs headshotKills heals killPlace killPoints kills killStreaks longestKill matchDuration matchType maxPlace numGroups rankPoints revives rideDistance roadKills swimDistance teamKills vehicleDestroys walkDistance weaponsAcquired winPoints winPlacePerc 7f96b2f878858a 4d4b580de459be a10357fd1a4a91 0 0 0.0 0 0 0 60 1241 0 0 0.0 1306 squad-fpp 28 26 -1 0 0.0 0 0.0 0 0 244.8 1 1466 0.4444 eef90569b9d03c 684d5656442f9e aeb375fc57110c 0 0 91.47 0 0 0 57 0 0 0 0.0 1777 squad-fpp 26 25 1484 0 0.0045 0 11.04 0 0 1434.0 5 0 0.64 1eaf90ac73de72 6a4a42c3245a74 110163d8bb94ae 1 0 68.0 0 0 0 47 0 0 0 0.0 1318 duo 50 47 1491 0 0.0 0 0.0 0 0 161.8 2 0 0.7755 4616d365dd2853 a930a9c79cd721 f1f1f4ef412d7e 0 0 32.9 0 0 0 75 0 0 0 0.0 1436 squad-fpp 31 30 1408 0 0.0 0 0.0 0 0 202.7 3 0 0.1667 315c96c26c9aac de04010b3458dd 6dc8ff871e21e6 0 0 100.0 0 0 0 45 0 1 1 58.53 1424 solo-fpp 97 95 1560 0 0.0 0 0.0 0 0 49.75 2 0 0.1875 ff79c12f326506 289a6836a88d27 bac52627a12114 0 0 100.0 1 1 0 44 0 1 1 18.44 1395 squad-fpp 28 28 1418 0 0.0 0 0.0 0 0 34.7 1 0 0.037 95959be0e21ca3 2c485a1ad3d0f1 a8274e903927a2 0 0 0.0 0 0 0 96 1262 0 0 0.0 1316 squad-fpp 28 28 -1 0 0.0 0 0.0 0 0 13.5 1 1497 0.0 311b84c6ff4390 eaba5fcb7fc1ae 292611730ca862 0 0 8.538 0 0 0 48 1000 0 0 0.0 1967 solo-fpp 96 92 -1 0 2004.0 0 0.0 0 0 1089.0 6 1500 0.7368 1a68204ccf9891 47cfbb04e1b1a2 df014fbee741c6 0 0 51.6 0 0 0 64 0 0 0 0.0 1375 squad 28 27 1493 0 0.0 0 0.0 0 0 799.9 4 0 0.3704 e5bb5a43587253 759bb6f7514fd2 3d3031c795305b 0 0 37.27 0 0 0 74 0 0 0 0.0 1930 squad 29 27 1349 0 0.0 0 0.0 0 0 65.67 1 0 0.2143 2b574d43972813 c549efede67ad3 2dd6ddb8320fc1 0 0 28.38 0 0 0 75 0 0 0 0.0 1811 squad-fpp 29 29 1475 0 0.0 0 0.0 0 0 868.3 9 0 0.3929 8de328a74658a9 f643df9df3877c 80170383d90003 0 0 137.9 1 0 0 64 0 0 0 0.0 1384 duo-fpp 48 46 1488 0 0.0 0 0.0 0 0 451.7 1 0 0.4043 ce4f6ac165705e da24cdb91969cc 535b5dbd965a94 0 0 0.0 0 0 0 37 0 0 0 0.0 1774 squad-fpp 29 28 1766 0 6639.0 0 0.0 0 0 2784.0 6 0 0.9286 b7807186e3f679 3c08e461874749 2c30ddf481c52d 0 1 324.2 0 1 5 5 986 4 1 49.83 1886 solo-fpp 97 94 -1 0 1228.0 0 76.84 0 0 2050.0 6 1462 0.875 8e244ac61b6aab d40d0c7d3573a1 94e1c1cc443c65 0 1 122.8 1 0 2 25 1411 1 1 37.91 1458 squad-fpp 31 30 -1 1 1237.0 0 60.29 0 0 1666.0 5 1531 0.9 12d8d4bd94312c fe52d481bae68b 6fd9e765ddd0c5 0 0 80.71 1 0 0 72 0 0 0 0.0 1313 duo-fpp 48 48 1495 1 0.0 0 0.0 0 0 105.1 5 0 0.2766 62f2f0917d84b2 f61b698274d9f5 1d6cfe0f6f23b0 0 2 81.71 1 0 14 25 0 1 1 9.158 1882 squad-fpp 27 26 1436 0 519.9 0 0.0 0 0 3674.0 7 0 0.7308 92022479b92ce7 2f2c33f548c4b9 07948d723b9c0f 0 3 254.3 0 0 12 13 0 2 1 36.0 1371 solo-fpp 96 95 1536 0 2367.0 0 15.29 0 0 1787.0 3 0 0.8211 7bd224781f064b 6dde607d151819 733af30cc00099 0 0 0.0 0 0 0 79 0 0 0 0.0 1301 squad-fpp 27 26 1355 0 0.0 0 0.0 0 0 137.4 2 0 0.1923 71cbdbc3b263e5 7b61f74b51906c a329ac99449ad7 0 1 65.28 0 0 1 48 1349 0 0 0.0 1322 squad-fpp 30 28 0 0 0.0 0 20.85 0 0 3310.0 3 1479 0.931 02ace8c6e58461 a4bc548028f800 80f2b8448e474b 0 4 269.1 0 1 8 18 0 2 1 7.438 1890 duo-fpp 48 48 1491 1 2734.0 0 0.0 0 0 1794.0 5 0 0.6383 00341b1caa5420 d661a2d19e7ae9 f3956286eb39a5 0 0 158.7 1 0 0 75 0 0 0 0.0 1853 squad-fpp 29 28 1522 0 0.0 0 0.0 0 0 580.1 2 0 0.2143 9b2961d4d51f91 799d0a4d61dc3c e833ca2282169d 0 1 192.3 1 2 3 15 1022 2 1 280.6 1775 duo-fpp 49 48 -1 0 2332.0 0 0.0 0 0 1264.0 4 1494 0.75 0b6fbdfb59c994 7a75c3e86934f6 8b0a78c005cea0 0 6 1011.0 6 2 2 2 1564 9 2 186.6 1857 duo 50 45 -1 0 4860.0 0 0.0 0 0 2727.0 7 1603 0.9592 736eda9b9c20b3 d35e80e4e64dd4 62fbe726028662 0 3 327.6 4 1 1 3 1346 4 2 167.8 1454 squad-fpp 27 27 -1 0 0.0 0 0.0 0 0 3503.0 4 1496 0.9231 4c45dc732689ec 8e0a0ea95d3596 37f43ba55ec0a4 1 4 558.6 3 0 4 11 1265 3 2 237.5 1841 duo-fpp 47 45 -1 0 1183.0 0 0.0 0 0 2711.0 7 1494 0.8696 91f5da9c3628eb 090fd12f3ca8a8 e8b8cf5d9231d3 0 0 44.28 0 0 0 78 0 0 0 0.0 1879 squad-fpp 27 25 1576 0 0.0 0 0.0 0 0 15.3 0 0 0.1154 dbf611495bfda3 10cbb86844dee0 2cdae31ee18601 0 4 381.2 2 

In [0]:
# Apply the best model
resultsDF = dtModel.transform(testDF).select("Id", "Prediction_WP")
display(resultsDF)

Id Prediction_WP 7f96b2f878858a 0.44290817142857136 eef90569b9d03c 0.5481449619771863 1eaf90ac73de72 0.7499144223693892 4616d365dd2853 0.22581101910827986 315c96c26c9aac 0.15800073452256047 ff79c12f326506 0.057620225593154474 95959be0e21ca3 0.007425869164183345 311b84c6ff4390 0.6720893593919652 1a68204ccf9891 0.40462416829745596 e5bb5a43587253 0.21816365784499053 2b574d43972813 0.23608395833333334 8de328a74658a9 0.3778184434968017 ce4f6ac165705e 0.8757934616159585 b7807186e3f679 0.8618593147751606 8e244ac61b6aab 0.7198561764705881 12d8d4bd94312c 0.2546947280930896 62f2f0917d84b2 0.8085140839218946 92022479b92ce7 0.8406158003088011 7bd224781f064b 0.16772235015238754 71cbdbc3b263e5 0.805775968222443 02ace8c6e58461 0.7368622543950363 00341b1caa5420 0.2379411111111106 9b2961d4d51f91 0.6138675099866843 0b6fbdfb59c994 0.8791398081534769 736eda9b9c20b3 0.912053259871442 4c45dc732689ec 0.8285518072289155 91f5da9c3628eb 0.1714520210896309 dbf611495bfda3 0.7502427135678391 f9473c4f1cfdc4 0.9197161544523254 ac5b57ff39979c 0.011046614664586587 3a4e362d28f411 0.42600999231360503 4d16222292e9c8 0.15157934058794847 e050d1f179d289 0.4918509833585477 4b604176d374e0 0.4617117924528302 e2d4735f01b693 0.369495686900958 47143f942503e0 0.28669288389513115 269c3fc4a26935 0.7910524699599468 c520f86dcd195a 0.12048754869876871 5cb863f7ac0dcd 0.8599682673588579 6e55ca8db3570e 0.23555439320388344 ffd9e56f13438e 0.39967753333333317 5b0bb2ecf3c824 0.08344410122813552 ca3c89d8045645 0.057620225593154474 bc6194d5fa2fe5 0.40998614599686023 46ec2eef21c254 0.8062476313079303 921456e81f6fe7 0.7920884615384615 316fce57a4480c 0.19183756331279941 8dfd8c4ea2861d 0.16157670316301706 0ba4529562941c 0.42405818403908807 c5d63a04ebad0e 0.5619281907090463 c67405a29d8d94 0.4589072968490879 e36b06a7553224 0.3483046528803546 e375584f42732e 0.35313884816753927 6178abb3fa1d4c 0.18400439770554494 6d22bbb91be111 0.7783979797979798 d79cdeb0ac6eb5 0.867355168661589 ccd2eb49c09d77 0.858201680672269 7e8f5c11aa9dd9 0.4669862580326248 a02d62f153b033 0.11852713178294574 d551381427e157 0.6598688771665411 5c46a37a783839 0.8312441375076735 dd87e30dc17eff 0.9200973760932949 eea397f23204bd 0.06988926585094547 73348483a5974b 0.1790211570247933 d2690c6ee5798c 0.38237362045760437 b83ba80e820303 0.7186161929170555 43bfa2926267fc 0.9487432847275519 22252be49d4419 0.928454434371305 b980859d30a6b2 0.19549819987585346 caa1a36afeb7b2 0.38128490770901197 83fac43ac28f44 0.6363441451990633 83d0b9e08b6e72 0.8069060674157305 84aa77226db78f 0.8090289156626508 a6ee0f6ffa8bb6 0.9061987693953986 8b3cdbefd1f408 0.0363487414187643 03cab96432e075 0.6169666666666666 5bb54605cf686f 0.30936923076923073 5d8016d7aa9437 0.6672396825396826 b23e720058fc87 0.08757735250186703 306cdfb564445f 0.30141678571428565 9f3b1c12a9f2ec 0.03198904494382026 ee3025cc18e7c9 0.9069855405405409 ba6ae608ca5f37 0.29995419161676645 498c5db01d537c 0.5727641921397381 e62133b0c6f55d 0.5111517857142857 95aaab951e3eec 0.8142862684729064 07719d6195d86b 0.04719130723291313 3dc087431e40f1 0.7723875899280576 416fafd3666b44 0.48776701570680614 677089d4226510 0.13417852841057867 35d182d668c400 0.495081081081081 5fd62798396ca8 0.10665709104098661 278fb2f43fc27b 0.20029569892473095 48649b6932f7df 0.37419894179894175 e14843dc63bf7a 0.37314604731679174 7f3d7f85e7736e 0.622539656311963 707756a08b72fe 0.4300159858947986 2bea3a18442001 0.8040993920972646 8fc5c0d83de539 0.5548921787709497 249c7209d743ed 0.10398434370057989 18d002b46b1abc 0.8483908875739647 707550186fd7f7 0.8744360824742269 04c32b79188af7 0.10431575647249187 c4260ce3746cf8 0.02133330930064888 ea298bfc126b74 0.7800717238899876 9df583833b18fb 0.7458791443850268 cfa11f07d80695 0.6984984654731459 7970b4ea89e81e 0.9047950060024011 4eb922611c2099 0.2494015037593985 de93ba20446980 0.007425869164183345 de1e7870d95850 0.10261518105849582 be91dbb39a9248 0.5184091962905719 05a0034ba0b373 0.28083795826283114 e264f7f399aa6a 0.9087196396396398 caf2cdfae38ea0 0.2208601587008672 97826efa3ce126 0.518584114052